"use strict";(self.webpackChunkwebsite=self.webpackChunkwebsite||[]).push([[5387],{3905:function(e,t,n){n.d(t,{Zo:function(){return p},kt:function(){return m}});var a=n(7294);function r(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function i(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);t&&(a=a.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,a)}return n}function o(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?i(Object(n),!0).forEach((function(t){r(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):i(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function s(e,t){if(null==e)return{};var n,a,r=function(e,t){if(null==e)return{};var n,a,r={},i=Object.keys(e);for(a=0;a<i.length;a++)n=i[a],t.indexOf(n)>=0||(r[n]=e[n]);return r}(e,t);if(Object.getOwnPropertySymbols){var i=Object.getOwnPropertySymbols(e);for(a=0;a<i.length;a++)n=i[a],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(r[n]=e[n])}return r}var l=a.createContext({}),c=function(e){var t=a.useContext(l),n=t;return e&&(n="function"==typeof e?e(t):o(o({},t),e)),n},p=function(e){var t=c(e.components);return a.createElement(l.Provider,{value:t},e.children)},u={inlineCode:"code",wrapper:function(e){var t=e.children;return a.createElement(a.Fragment,{},t)}},d=a.forwardRef((function(e,t){var n=e.components,r=e.mdxType,i=e.originalType,l=e.parentName,p=s(e,["components","mdxType","originalType","parentName"]),d=c(n),m=r,h=d["".concat(l,".").concat(m)]||d[m]||u[m]||i;return n?a.createElement(h,o(o({ref:t},p),{},{components:n})):a.createElement(h,o({ref:t},p))}));function m(e,t){var n=arguments,r=t&&t.mdxType;if("string"==typeof e||r){var i=n.length,o=new Array(i);o[0]=d;var s={};for(var l in t)hasOwnProperty.call(t,l)&&(s[l]=t[l]);s.originalType=e,s.mdxType="string"==typeof e?e:r,o[1]=s;for(var c=2;c<i;c++)o[c]=n[c];return a.createElement.apply(null,o)}return a.createElement.apply(null,n)}d.displayName="MDXCreateElement"},9162:function(e,t,n){n.r(t),n.d(t,{default:function(){return u},frontMatter:function(){return s},metadata:function(){return l},toc:function(){return c}});var a=n(7462),r=n(3366),i=(n(7294),n(3905)),o=["components"],s={id:"publications",title:"Publications"},l={unversionedId:"publications",id:"publications",isDocsHomePage:!1,title:"Publications",description:"Here are the current papers associated with Octopus.",source:"@site/docs/publications.md",sourceDirName:".",slug:"/publications",permalink:"/octopus/docs/publications",editUrl:"https://github.com/${organizationName}/${projectName}/edit/${branch}/website/docs/publications.md",version:"current",frontMatter:{id:"publications",title:"Publications"}},c=[{value:"A unified haplotype-based method for accurate and comprehensive variant calling",id:"a-unified-haplotype-based-method-for-accurate-and-comprehensive-variant-calling",children:[]},{value:"Benchmarking small-variant genotyping in polyploids",id:"benchmarking-small-variant-genotyping-in-polyploids",children:[]},{value:"Accurate genotyping of single cells with Octopus",id:"accurate-genotyping-of-single-cells-with-octopus",children:[]}],p={toc:c};function u(e){var t=e.components,n=(0,r.Z)(e,o);return(0,i.kt)("wrapper",(0,a.Z)({},p,n,{components:t,mdxType:"MDXLayout"}),(0,i.kt)("p",null,"Here are the current papers associated with Octopus."),(0,i.kt)("h2",{id:"a-unified-haplotype-based-method-for-accurate-and-comprehensive-variant-calling"},(0,i.kt)("a",{parentName:"h2",href:"https://www.nature.com/articles/s41587-021-00861-3"},"A unified haplotype-based method for accurate and comprehensive variant calling")),(0,i.kt)("p",null,"This is the main Octopus paper describing the method with benchmarks for germline and somatic variant calling."),(0,i.kt)("h4",{id:"abstract"},"Abstract"),(0,i.kt)("p",null,"Almost all haplotype-based variant callers were designed specifically for detecting common germline variation in diploid populations, and give suboptimal results in other scenarios. Here we present Octopus, a variant caller that uses a polymorphic Bayesian genotyping model capable of modeling sequencing data from a range of experimental designs within a unified haplotype-aware framework. Octopus combines sequencing reads and prior information to phase-called genotypes of arbitrary ploidy, including those with somatic mutations. We show that Octopus accurately calls germline variants in individuals, including single nucleotide variants, indels and small complex replacements such as microinversions. Using a synthetic tumor data set derived from clean sequencing data from a sample with known germline haplotypes and observed mutations in a large cohort of tumor samples, we show that Octopus is more sensitive to low-frequency somatic variation, yet calls considerably fewer false positives than other methods. Octopus also outputs realigned evidence BAM files to aid validation and interpretation."),(0,i.kt)("h4",{id:"bibtex"},"Bibtex"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-tex"},"@article{octopus,\n   author = {Cooke, Daniel P. and Wedge, David C. and Lunter, Gerton},\n   title = {A unified haplotype-based method for accurate and comprehensive variant calling},\n   journal = {Nature Biotechnology},\n   ISSN = {1546-1696},\n   DOI = {10.1038/s41587-021-00861-3},\n   url = {https://doi.org/10.1038/s41587-021-00861-3\nhttps://www.nature.com/articles/s41587-021-00861-3.pdf},\n   year = {2021},\n   type = {Journal Article}\n}\n")),(0,i.kt)("h2",{id:"benchmarking-small-variant-genotyping-in-polyploids"},(0,i.kt)("a",{parentName:"h2",href:"https://genome.cshlp.org/content/32/2/403"},"Benchmarking small-variant genotyping in polyploids")),(0,i.kt)("p",null,"In this paper, we benchmark Octopus on polyploid samples."),(0,i.kt)("h4",{id:"abstract-1"},"Abstract"),(0,i.kt)("p",null,"Genotyping from sequencing is the basis of emerging strategies in the molecular breeding of polyploid plants. However, compared with the situation for diploids, in which genotyping accuracies are confidently determined with comprehensive benchmarks, polyploids have been neglected; there are no benchmarks measuring genotyping error rates for small variants using real sequencing reads. We previously introduced a variant calling method, Octopus, that accurately calls germline variants in diploids and somatic mutations in tumors. Here, we evaluate Octopus and other popular tools on whole-genome tetraploid and hexaploid data sets created using in silico mixtures of diploid Genome in a Bottle (GIAB) samples. We find that genotyping errors are abundant for typical sequencing depths but that Octopus makes 25% fewer errors than other methods on average. We supplement our benchmarks with concordance analysis in real autotriploid banana data sets."),(0,i.kt)("h4",{id:"bibtex-1"},"Bibtex"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-tex"},"@article{Cooke01022022,\nauthor = {Cooke, Daniel P. and Wedge, David C. and Lunter, Gerton}, \ntitle = {Benchmarking small-variant genotyping in polyploids},\nvolume = {32}, \nnumber = {2}, \npages = {403-408}, \nyear = {2022}, \ndoi = {10.1101/gr.275579.121}, \nURL = {http://genome.cshlp.org/content/32/2/403.abstract}, \neprint = {http://genome.cshlp.org/content/32/2/403.full.pdf+html}, \njournal = {Genome Research} \n}\n")),(0,i.kt)("h2",{id:"accurate-genotyping-of-single-cells-with-octopus"},(0,i.kt)("a",{parentName:"h2",href:"https://www.researchsquare.com/article/rs-583831/v1"},"Accurate genotyping of single cells with Octopus")),(0,i.kt)("h4",{id:"absract"},"Absract"),(0,i.kt)("p",null,"We describe an extension to our variant calling tool, Octopus (",(0,i.kt)("a",{parentName:"p",href:"https://github.com/luntergroup/octopus"},"https://github.com/luntergroup/octopus"),"), for single-cell DNA sequencing data. Octopus jointly genotypes cells from a lineage, accounting for amplification stochasticity and sequencing error with a haplotype-based Bayesian model. Octopus is considerably more accurate at genotyping single cells than existing methods."),(0,i.kt)("h4",{id:"bibtex-2"},"Bibtex"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre",className:"language-tex"},"@article{RN862,\n   author = {Daniel, Cooke P. and Gerton, Lunter and David C., Wedge},\n   title = {Accurate genotyping of single cells with Octopus},\n   journal = {Research Square},\n   ISSN = {2693-5015},\n   DOI = {10.21203/rs.3.rs-583831/v1},\n   url = {https://doi.org/10.21203/rs.3.rs-583831/v1},\n   year = {2021},\n   type = {Journal Article}\n}\n\n")))}u.isMDXComponent=!0}}]);