<!doctype html>
<html lang="en" dir="ltr">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.0">
<link rel="alternate" type="application/rss+xml" href="/octopus/blog/rss.xml" title="Octopus Blog RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/octopus/blog/atom.xml" title="Octopus Blog Atom Feed"><title data-react-helmet="true">Population | Octopus</title><meta data-react-helmet="true" property="og:url" content="https://luntergroup.github.io/octopus/docs/guides/models/population"><meta data-react-helmet="true" name="docusaurus_locale" content="en"><meta data-react-helmet="true" name="docusaurus_version" content="current"><meta data-react-helmet="true" name="docusaurus_tag" content="docs-default-current"><meta data-react-helmet="true" property="og:title" content="Population | Octopus"><meta data-react-helmet="true" name="description" content="The population calling model is for jointly calling germline variation in small cohorts of samples with known ploidy but unknown pedigree structure."><meta data-react-helmet="true" property="og:description" content="The population calling model is for jointly calling germline variation in small cohorts of samples with known ploidy but unknown pedigree structure."><link data-react-helmet="true" rel="shortcut icon" href="/octopus/img/favicon.ico"><link data-react-helmet="true" rel="canonical" href="https://luntergroup.github.io/octopus/docs/guides/models/population"><link data-react-helmet="true" rel="alternate" href="https://luntergroup.github.io/octopus/docs/guides/models/population" hreflang="en"><link data-react-helmet="true" rel="alternate" href="https://luntergroup.github.io/octopus/docs/guides/models/population" hreflang="x-default"><link rel="stylesheet" href="/octopus/assets/css/styles.0d769f7d.css">
<link rel="preload" href="/octopus/assets/js/runtime~main.3fccafd2.js" as="script">
<link rel="preload" href="/octopus/assets/js/main.f8c4a271.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div><a href="#main" class="skipToContent_OuoZ">Skip to main content</a></div><nav class="navbar navbar--fixed-top navbarHideable_RReh"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle" type="button" tabindex="0"><svg aria-label="Menu" width="30" height="30" viewBox="0 0 30 30" role="img" focusable="false"><title>Menu</title><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/octopus/"><img src="/octopus/img/octopus.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--light_4Vu1 navbar__logo"><img src="/octopus/img/octopus_keytar.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--dark_uzRr navbar__logo"><strong class="navbar__title">Octopus</strong></a><a class="navbar__item navbar__link navbar__link--active" href="/octopus/docs/introduction">Docs</a><a class="navbar__item navbar__link" href="/octopus/docs/cli">Command line</a><a class="navbar__item navbar__link" href="/octopus/docs/publications">Publications</a><a class="navbar__item navbar__link" href="/octopus/blog">Blog</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/luntergroup/octopus" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="react-toggle displayOnlyInLargeViewport_cxYs react-toggle--disabled" role="button" tabindex="-1"><div class="react-toggle-track"><div class="react-toggle-track-check"><span class="toggle_iYfV">ðŸŒœ</span></div><div class="react-toggle-track-x"><span class="toggle_iYfV">ðŸŒž</span></div></div><div class="react-toggle-thumb"></div><input type="checkbox" class="react-toggle-screenreader-only" aria-label="Switch between dark and light mode"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div><div class="navbar-sidebar"><div class="navbar-sidebar__brand"><a class="navbar__brand" href="/octopus/"><img src="/octopus/img/octopus.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--light_4Vu1 navbar__logo"><img src="/octopus/img/octopus_keytar.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--dark_uzRr navbar__logo"><strong class="navbar__title">Octopus</strong></a></div><div class="navbar-sidebar__items"><div class="menu"><ul class="menu__list"><li class="menu__list-item"><a class="menu__link navbar__link--active" href="/octopus/docs/introduction">Docs</a></li><li class="menu__list-item"><a class="menu__link" href="/octopus/docs/cli">Command line</a></li><li class="menu__list-item"><a class="menu__link" href="/octopus/docs/publications">Publications</a></li><li class="menu__list-item"><a class="menu__link" href="/octopus/blog">Blog</a></li><li class="menu__list-item"><a href="https://github.com/luntergroup/octopus" target="_blank" rel="noopener noreferrer" class="menu__link header-github-link" aria-label="GitHub repository"></a></li></ul></div></div></div></nav><div class="main-wrapper docs-wrapper doc-page"><div class="docPage_lDyR"><div class="docSidebarContainer_0YBq" role="complementary"><div class="sidebar_a3j0 sidebarWithHideableNavbar_VlPv"><a tabindex="-1" class="sidebarLogo_hmkv" href="/octopus/"><img src="/octopus/img/octopus.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--light_4Vu1"><img src="/octopus/img/octopus_keytar.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--dark_uzRr"><strong>Octopus</strong></a><div class="menu menu--responsive thin-scrollbar menu_cyFh"><button aria-label="Open menu" aria-haspopup="true" class="button button--secondary button--sm menu__button" type="button"><svg aria-label="Menu" class="sidebarMenuIcon_iZzd" width="24" height="24" viewBox="0 0 30 30" role="img" focusable="false"><title>Menu</title><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><ul class="menu__list"><li class="menu__list-item"><a class="menu__link menu__link--sublist" href="#!">Getting Started</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/introduction">Introduction</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/installation">Installation</a></li></ul></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#!">Guides</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/overview">Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/preprocessing">Read Preprocessing</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/discovery">Variant Discovery</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/haplotypes">Haplotype Proposal</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#!" tabindex="0">Calling Models</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/models/individual">Individual</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/models/trio">Trio</a></li><li class="menu__list-item"><a aria-current="page" class="menu__link menu__link--active active" tabindex="0" href="/octopus/docs/guides/models/population">Population</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/models/cancer">Cancer</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/models/polyclone">Polyclone</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/models/cell">Cell</a></li></ul></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/errorModels">Error Models</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/phasing">Phasing</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">Filtering</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/filtering/annotations">Introduction</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/filtering/forest">Random Forest</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/filtering/thresholds">Hard Thresholds</a></li></ul></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/bamout">Realigned BAMs</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">Advanced</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/targeted">Targeted Calling</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/threading">Multithreading</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/memory">Memory Use</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/vcf">VCF Format</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/configs">Config files</a></li></ul></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!">Case studies</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/tutorials/germline">Germline WGS</a></li></ul></li></ul></div></div></div><main class="docMainContainer_r8cw"><div class="container padding-vert--lg docItemWrapper_NJLN"><div class="row"><div class="col docItemCol_zHA2"><div class="docItemContainer_oiyr"><article><header><h1 class="docTitle_-X99">Population</h1></header><div class="markdown"><p>The <code>population</code> calling model is for jointly calling germline variation in small cohorts of samples with known ploidy but unknown pedigree structure. </p><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="usage"></a>Usage<a class="hash-link" href="#usage" title="Direct link to heading">#</a></h2><p>Multiple samples from the same population, without pedigree information, can be called jointly:</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R hs37d5.fa -I NA12878.bam NA12891.bam NA12892.bam</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p>Joint calling samples may increase calling power, especially for low coverage sequencing.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="setting-ploidies"></a>Setting ploidies<a class="hash-link" href="#setting-ploidies" title="Direct link to heading">#</a></h2><p>Octopus can currently joint call samples that have the same ploidies for all contigs. You can set ploidies with the <code>--organism-ploidy</code> (<code>-P</code>):</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R hs37d5.fa -I haploid1.bam haploid2.bam haploid3.bam -P </span><span class="token number" style="color:rgb(247, 140, 108)">1</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p>and <code>--contig-ploidies</code> (<code>-p</code>) options</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R hs37d5.fa -I NA12878.bam NA12891.bam NA12892.bam -p </span><span class="token assign-left variable" style="color:rgb(191, 199, 213)">X</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token number" style="color:rgb(247, 140, 108)">1</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p>Octopus sets the organism ploidy to 2 by default and also sets chromosome Y to ploidy 1. X is left as two, so if your samples are all male, you could set <code>-p X=1</code>.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="performance-considerations"></a>Performance considerations<a class="hash-link" href="#performance-considerations" title="Direct link to heading">#</a></h2><p>For joint calling, the single most important parameter which will determine performance is <a href="/octopus/docs/cli#--max-genotype-combinations"><code>--max-genotype-combinations</code></a>. This parameter determines the maximum number of joint genotype combinations that octopus can consider. The number of possibl genotype combinations at a given site is exponential in the number of samples, so this value will usually be significantly lower than the number of possible joint genotypes. Octopus will select the &#x27;best&#x27; <code>--max-genotype-combinations</code> genotype combinations using approximation. Increasing this value may improve accuracy, but result in longer runtimes and memory use.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="calling-large-cohorts-and-the-n--1-problem"></a>Calling large cohorts, and the n + 1 problem<a class="hash-link" href="#calling-large-cohorts-and-the-n--1-problem" title="Direct link to heading">#</a></h2><p>Joint calling many samples (&gt; 10), especially high coverage ones, is computationally expensive. This is because true joint calling necessarily requires considering all possible genotypes for each sample, which grow approximately linearly with the number of samples (due to noise). The number of possible genotype combinations is also polynomial in the number of possible genotypes.</p><p>There is also the so-called &#x27;n + 1 problem&#x27; of adding an extra sample to a cohort already jointly called. Ideally this would be done cheaply rather than having to recall all samples again. Other methods have addressed this problem by individually calling all samples, then jointly calling all samples by evaluating previously computed genotype likelihoods under a joint model. However, this approach will never correctly call cases where a true allele was not proposed for a specific individual (since no genotype likelihood will be computed). This method also suffers the same &#x27;representation problems&#x27; as typical merge procedures as it does not consider haplotypes.</p><p>Although octopus does not specifically address the n + 1 problem, we suggest a procedure for jointly calling many samples:</p><ol><li>Call each sample individually.</li><li>Group the samples into small subsets (&lt; 20), joint call these using only the variants called previously.</li><li>Repeat step 2 until there is only one group left (containing all samples).</li></ol><p>Adding another sample is achieved in a similar way:</p><ol><li>Call the new sample individually.</li><li>Joint call all samples using the variants called in previous joint calling and the new sample.</li></ol><p>The reason these procedures is cheaper than joint calling de-novo is the number of candidate alleles is likely significantly reduced as most noise should be removed during individual calling.</p><p>For example, three samples could be jointly called with this procedure as follows:</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R hs37d5.fa -I NA12878.bam -o octopus.NA12878.bcf</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R hs37d5.fa -I NA12891.bam -o octopus.NA12891.bcf</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R hs37d5.fa -I NA12892.bam -o octopus.NA12892.bcf</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R hs37d5.fa -I NA12892.bam NA12891.bam NA12892.bam </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"> --disable-denovo-variant-discovery </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"> -c octopus.NA12878.bcf octopus.NA12891.bcf octopus.NA12892.bcf </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"> -o octopus.joint.bcf</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p>Note the last step disables candidate generation from raw alignments (<code>-g</code>) and local reassembly (<code>-a</code>), and only uses provided source variants (<code>-c</code>).</p></div></article><div class="margin-vert--xl"><div class="row"><div class="col"><a href="https://github.com/${organizationName}/${projectName}/edit/${branch}/website/docs/guides/models/population.md" target="_blank" rel="noreferrer noopener"><svg fill="currentColor" height="1.2em" width="1.2em" preserveAspectRatio="xMidYMid meet" role="img" viewBox="0 0 40 40" class="iconEdit_mS5F" aria-label="Edit page"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div></div></div><div class="margin-vert--lg"><nav class="pagination-nav" aria-label="Docs pages navigation"><div class="pagination-nav__item"><a class="pagination-nav__link" href="/octopus/docs/guides/models/trio"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Â« Trio</div></a></div><div class="pagination-nav__item pagination-nav__item--next"><a class="pagination-nav__link" href="/octopus/docs/guides/models/cancer"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Cancer Â»</div></a></div></nav></div></div></div><div class="col col--3"><div class="tableOfContents_vrFS thin-scrollbar"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#usage" class="table-of-contents__link">Usage</a></li><li><a href="#setting-ploidies" class="table-of-contents__link">Setting ploidies</a></li><li><a href="#performance-considerations" class="table-of-contents__link">Performance considerations</a></li><li><a href="#calling-large-cohorts-and-the-n--1-problem" class="table-of-contents__link">Calling large cohorts, and the n + 1 problem</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><h4 class="footer__title">Docs</h4><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/octopus/docs/installation">Installation</a></li><li class="footer__item"><a class="footer__link-item" href="/octopus/docs/cli">Command Line</a></li><li class="footer__item"><a class="footer__link-item" href="/octopus/docs/tutorials/germline">Case Studies</a></li></ul></div><div class="col footer__col"><h4 class="footer__title">Community</h4><ul class="footer__items"><li class="footer__item"><a href="https://www.biostars.org/tag/octopus" target="_blank" rel="noopener noreferrer" class="footer__link-item">Biostars</a></li><li class="footer__item"><a href="https://gitter.im/octopus-caller/Lobby" target="_blank" rel="noopener noreferrer" class="footer__link-item">Gitter</a></li><li class="footer__item"><a href="https://twitter.com/octopus" target="_blank" rel="noopener noreferrer" class="footer__link-item">Twitter</a></li></ul></div><div class="col footer__col"><h4 class="footer__title">More</h4><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/octopus/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/luntergroup/octopus" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub</a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright Â© 2022 Daniel Cooke. Built with Docusaurus.</div></div></div></footer></div>
<script src="/octopus/assets/js/runtime~main.3fccafd2.js"></script>
<script src="/octopus/assets/js/main.f8c4a271.js"></script>
</body>
</html>