<!doctype html>
<html lang="en" dir="ltr">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.0">
<link rel="alternate" type="application/rss+xml" href="/octopus/blog/rss.xml" title="Octopus Blog RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/octopus/blog/atom.xml" title="Octopus Blog Atom Feed"><title data-react-helmet="true">Command Line Reference | Octopus</title><meta data-react-helmet="true" property="og:url" content="https://luntergroup.github.io/octopus/docs/cli"><meta data-react-helmet="true" name="docusaurus_locale" content="en"><meta data-react-helmet="true" name="docusaurus_version" content="current"><meta data-react-helmet="true" name="docusaurus_tag" content="docs-default-current"><meta data-react-helmet="true" property="og:title" content="Command Line Reference | Octopus"><meta data-react-helmet="true" name="description" content="General options"><meta data-react-helmet="true" property="og:description" content="General options"><link data-react-helmet="true" rel="shortcut icon" href="/octopus/img/favicon.ico"><link data-react-helmet="true" rel="canonical" href="https://luntergroup.github.io/octopus/docs/cli"><link data-react-helmet="true" rel="alternate" href="https://luntergroup.github.io/octopus/docs/cli" hreflang="en"><link data-react-helmet="true" rel="alternate" href="https://luntergroup.github.io/octopus/docs/cli" hreflang="x-default"><link rel="stylesheet" href="/octopus/assets/css/styles.0d769f7d.css">
<link rel="preload" href="/octopus/assets/js/runtime~main.e4a68a25.js" as="script">
<link rel="preload" href="/octopus/assets/js/main.f8c4a271.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div><a href="#main" class="skipToContent_OuoZ">Skip to main content</a></div><nav class="navbar navbar--fixed-top navbarHideable_RReh"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle" type="button" tabindex="0"><svg aria-label="Menu" width="30" height="30" viewBox="0 0 30 30" role="img" focusable="false"><title>Menu</title><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/octopus/"><img src="/octopus/img/octopus.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--light_4Vu1 navbar__logo"><img src="/octopus/img/octopus_keytar.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--dark_uzRr navbar__logo"><strong class="navbar__title">Octopus</strong></a><a class="navbar__item navbar__link" href="/octopus/docs/introduction">Docs</a><a aria-current="page" class="navbar__item navbar__link navbar__link--active navbar__link--active" href="/octopus/docs/cli">Command line</a><a class="navbar__item navbar__link navbar__link--active" href="/octopus/docs/publications">Publications</a><a class="navbar__item navbar__link" href="/octopus/blog">Blog</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/luntergroup/octopus" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="react-toggle displayOnlyInLargeViewport_cxYs react-toggle--disabled" role="button" tabindex="-1"><div class="react-toggle-track"><div class="react-toggle-track-check"><span class="toggle_iYfV">ðŸŒœ</span></div><div class="react-toggle-track-x"><span class="toggle_iYfV">ðŸŒž</span></div></div><div class="react-toggle-thumb"></div><input type="checkbox" class="react-toggle-screenreader-only" aria-label="Switch between dark and light mode"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div><div class="navbar-sidebar"><div class="navbar-sidebar__brand"><a class="navbar__brand" href="/octopus/"><img src="/octopus/img/octopus.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--light_4Vu1 navbar__logo"><img src="/octopus/img/octopus_keytar.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--dark_uzRr navbar__logo"><strong class="navbar__title">Octopus</strong></a></div><div class="navbar-sidebar__items"><div class="menu"><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" href="/octopus/docs/introduction">Docs</a></li><li class="menu__list-item"><a aria-current="page" class="menu__link navbar__link--active navbar__link--active" href="/octopus/docs/cli">Command line</a></li><li class="menu__list-item"><a class="menu__link navbar__link--active" href="/octopus/docs/publications">Publications</a></li><li class="menu__list-item"><a class="menu__link" href="/octopus/blog">Blog</a></li><li class="menu__list-item"><a href="https://github.com/luntergroup/octopus" target="_blank" rel="noopener noreferrer" class="menu__link header-github-link" aria-label="GitHub repository"></a></li></ul></div></div></div></nav><div class="main-wrapper docs-wrapper doc-page"><div class="docPage_lDyR"><main class="docMainContainer_r8cw docMainContainerEnhanced_SOUu"><div class="container padding-vert--lg docItemWrapper_NJLN"><div class="row"><div class="col docItemCol_zHA2"><div class="docItemContainer_oiyr"><article><header><h1 class="docTitle_-X99">Command Line Reference</h1></header><div class="markdown"><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="general-options"></a>General options<a class="hash-link" href="#general-options" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--help"></a><code>--help</code><a class="hash-link" href="#--help" title="Direct link to heading">#</a></h3><p>Command <code>--help</code> (short <code>-h</code>) prints the list of available command line options.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus --help</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">Octopus </span><span class="token builtin class-name" style="color:rgb(255, 203, 107)">command</span><span class="token plain"> line options:</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain" style="display:inline-block">
</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">General:</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">  -h </span><span class="token punctuation" style="color:rgb(199, 146, 234)">[</span><span class="token plain"> --help </span><span class="token punctuation" style="color:rgb(199, 146, 234)">]</span><span class="token plain">                         Report detailed option information</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">  --version                             Report detailed version information</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">  --config arg                          Config </span><span class="token function" style="color:rgb(130, 170, 255)">file</span><span class="token plain"> to populate </span><span class="token builtin class-name" style="color:rgb(255, 203, 107)">command</span><span class="token plain"> line </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">                                        options</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">  --options                             Log all </span><span class="token builtin class-name" style="color:rgb(255, 203, 107)">command</span><span class="token plain"> line option values at </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">                                        startup</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">  --debug </span><span class="token punctuation" style="color:rgb(199, 146, 234)">[</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token plain">arg</span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token string" style="color:rgb(195, 232, 141)">&quot;octopus_debug.log&quot;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token punctuation" style="color:rgb(199, 146, 234)">]</span><span class="token plain">  Create log </span><span class="token function" style="color:rgb(130, 170, 255)">file</span><span class="token plain"> </span><span class="token keyword" style="font-style:italic">for</span><span class="token plain"> debugging</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">  --trace </span><span class="token punctuation" style="color:rgb(199, 146, 234)">[</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token plain">arg</span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token string" style="color:rgb(195, 232, 141)">&quot;octopus_trace.log&quot;</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token punctuation" style="color:rgb(199, 146, 234)">]</span><span class="token plain">  Create very verbose log </span><span class="token function" style="color:rgb(130, 170, 255)">file</span><span class="token plain"> </span><span class="token keyword" style="font-style:italic">for</span><span class="token plain"> </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">                                        debugging</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">  -w </span><span class="token punctuation" style="color:rgb(199, 146, 234)">[</span><span class="token plain"> --working-directory </span><span class="token punctuation" style="color:rgb(199, 146, 234)">]</span><span class="token plain"> arg        Sets the working directory</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">  --threads </span><span class="token punctuation" style="color:rgb(199, 146, 234)">[</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token plain">arg</span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token number" style="color:rgb(247, 140, 108)">0</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token punctuation" style="color:rgb(199, 146, 234)">]</span><span class="token plain">                  Maximum number of threads to be used. </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">                                        If no argument is provided unlimited </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">                                        threads are assumed</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&lt;</span><span class="token plain">OUTPUT STRIPPED</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--version"></a><code>--version</code><a class="hash-link" href="#--version" title="Direct link to heading">#</a></h3><p>Command <code>--version</code> prints the version of the current binary, including the git branch and commit. The command also prints some system information about the binary.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus --version</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">octopus version </span><span class="token number" style="color:rgb(247, 140, 108)">0.6</span><span class="token plain">.3-beta </span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token plain">develop e7205439</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">Target: x86_64 Darwin </span><span class="token number" style="color:rgb(247, 140, 108)">18.7</span><span class="token plain">.0</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">SIMD extension: AVX2</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">Compiler: AppleClang </span><span class="token number" style="color:rgb(247, 140, 108)">10.0</span><span class="token plain">.1.10010046</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">Boost: 1_70</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--config"></a><code>--config</code><a class="hash-link" href="#--config" title="Direct link to heading">#</a></h3><p>Option <code>--config</code> is used to specify <a href="/octopus/docs/guides/advanced/configs">configuration files</a>.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus ---config myconfig.txt</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--trace"></a><code>--trace</code><a class="hash-link" href="#--trace" title="Direct link to heading">#</a></h3><p>Option <code>--trace</code> results in a log file containing detailed debug information being created.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --trace </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># default name is &quot;octopus_trace.log&quot;</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --trace </span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token plain">/foo/bar/trace.log</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Warning</strong> Trace files can get very large, so only use on small inputs..</p><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--working-directory"></a><code>--working-directory</code><a class="hash-link" href="#--working-directory" title="Direct link to heading">#</a></h3><p>Option <code>--working-directory</code> (short <code>-w</code>) is used to set the working directory of the run. All output and temporary files will be relative to the working directory, unless absolute file paths are provided.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -w ~/vcf -o octopus.vcf</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p>The output will be in <code>~/vcf/octopus.vcf</code>.</p><p><strong>Notes</strong></p><ul><li>If no working directory is specified then the current directory is used.</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--resolve-symlinks"></a><code>--resolve-symlinks</code><a class="hash-link" href="#--resolve-symlinks" title="Direct link to heading">#</a></h3><p>Command <code>--resolve-symlinks</code> forces all symbolic link input paths to be replaced with their resolved targets during program initialisation.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --resolve-symlinks</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--threads"></a><code>--threads</code><a class="hash-link" href="#--threads" title="Direct link to heading">#</a></h3><p>Option <code>--threads</code> is used to enable <a href="/octopus/docs/guides/advanced/threading">multithreaded execution</a>. The option accepts a positive integer argument that specifies the maximum number of threads to use:</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --threads </span><span class="token number" style="color:rgb(247, 140, 108)">4</span><span class="token plain"> </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># use maximum of 4 threads</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p>If no argument is provided then automatic thread handling is used. This may be greater than the number of system threads.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --threads </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic">#automatic thread handling</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-reference-cache-memory"></a><code>--max-reference-cache-memory</code><a class="hash-link" href="#--max-reference-cache-memory" title="Direct link to heading">#</a></h3><p>Option <code>--max-reference-cache-memory</code> (short <code>-X</code>) controls the size of the buffer used for reference caching, and is therefore one way to <a href="/octopus/docs/guides/advanced/memory">control memory use</a>. The option accepts a non-negative integer argument in bytes, and an optional unit specifier.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -X </span><span class="token number" style="color:rgb(247, 140, 108)">0</span><span class="token plain"> </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># disable reference caching</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -X 1g </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># reference cache is 1 gigabyte</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>Capitisation of the units is ignored.</li><li>An argument of <code>0</code> disables reference caching.</li><li>The buffer size never exceeds the size of the reference.</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--target-read-buffer-memory"></a><code>--target-read-buffer-memory</code><a class="hash-link" href="#--target-read-buffer-memory" title="Direct link to heading">#</a></h3><p>Option <code>--target-read-buffer-memory</code> (short <code>-B</code>) controls the size of the memory buffer used for input sequencing reads, and is therefore one way to <a href="/octopus/docs/guides/advanced/memory">control memory use</a>. The option accepts a positive integer argument in bytes, and an optional unit specifier.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -B 20G </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># reference cache is 20 gigabyte</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>Capitisation of the units is ignored.</li><li>The minimum read buffer size is <code>50Mb</code>; arguments less than this are ignored.</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--target-working-memory"></a><code>--target-working-memory</code><a class="hash-link" href="#--target-working-memory" title="Direct link to heading">#</a></h3><p>Option <code>--target-working-memory</code> sets the target amount of working memory for computation, and is therefore one way to <a href="/octopus/docs/guides/advanced/memory">control memory use</a>. The option accepts a positive integer argument in bytes, and an optional unit specifier. The option is not strictly enforced, but is sometimes used to decide whether to switch to lower-memory versions of some methods (possibly at the cost of additional runtime).</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --target-working-memory 40Gb</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>Capitisation of the units is ignored.</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--temp-directory-prefix"></a><code>--temp-directory-prefix</code><a class="hash-link" href="#--temp-directory-prefix" title="Direct link to heading">#</a></h3><p>Option <code>--temp-directory-prefix</code> sets the Octopus working temporary directory prefix.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --temp-directory-prefix ~/octopus_tmp</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>Like all path arguments in Octopus, the argument is assumed to be relative to the <code>--working-directory</code>, unless the path is absolute and exists.</li><li>If a directory with the prefix already exists, then <code>-N</code> is appended to the prefix where <code>N</code> is the smallest integer such that the resulting path does not exist.</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--reference"></a><code>--reference</code><a class="hash-link" href="#--reference" title="Direct link to heading">#</a></h3><p>Option <code>--reference</code> (short <code>-R</code>) sets the reference FASTA file used for variant calling. This is a required option.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>In addition to the FASTA file, a reference FASTA index file (extension <code>.fai</code>) is also required. This must be present in the same directory as the FASTA file.</li><li>Like all path arguments in Octopus, the argument is assumed to be relative to the <code>--working-directory</code>, unless the path is absolute and exists.</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--reads"></a><code>--reads</code><a class="hash-link" href="#--reads" title="Direct link to heading">#</a></h3><p>Option <code>--reads</code> (short <code>-I</code>) specifies the input read files (extension <code>.bam</code> or <code>.cram</code>) used for variant calling. The argument is a list of file paths. This is a required option (unless <code>--reads-file</code> is specified).</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads1.bam reads2.bam</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>Each read file must have a paired index file in the same directory.</li><li>The option can be specified multiple times on the command line (arguments are concatenated).</li><li>Can be used in conjunction with <code>--reads-file</code> (arguments are concatenated).</li><li>Like all path arguments in Octopus, the argument is assumed to be relative to the <code>--working-directory</code>, unless the path is absolute and exists.</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--reads-file"></a><code>--reads-file</code><a class="hash-link" href="#--reads-file" title="Direct link to heading">#</a></h3><p>Option <code>--reads-file</code> (short <code>-i</code>) specifies a file containing a list of input read files (one per line).</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -i bams.txt</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--regions"></a><code>--regions</code><a class="hash-link" href="#--regions" title="Direct link to heading">#</a></h3><p>Option <code>--regions</code> (short <code>-T</code>) specifies a list of genomic <a href="/octopus/docs/guides/advanced/targeted">regions to call</a>.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -T chr1 </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># all of chr1</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -T chr1 chr2 </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># all of chr1 and chr2</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -T chr1:0-1,000 </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># first 1,000bp of chr1</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -T chr1:100- </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># everything after position 100 of chr1</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -T chr1 to chrM </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># as specified in reference index</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>If this option is not specified (and neither is <code>--regions-file</code>), then all contigs present in the reference genome index are used.</li><li>Commas in the position tokens are ignored.</li><li>The option can be specified multiple times on the command line (arguments are concatenated).</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--regions-file"></a><code>--regions-file</code><a class="hash-link" href="#--regions-file" title="Direct link to heading">#</a></h3><p>Option <code>--regions-file</code> (short <code>-t</code>) specifies a file containing a list of genomic regions to call (one per line). The file format can either be plain text (using the same input format as the <code>--regions</code> option), or BED format (i.e. tab-separated). GZipped files are also accepted.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -t regions.bed</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--skip-regions"></a><code>--skip-regions</code><a class="hash-link" href="#--skip-regions" title="Direct link to heading">#</a></h3><p>Option <code>--skip-regions</code> (short <code>-K</code>) specifies a list of genomic regions to ignore during calling.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -K X Y MT</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>The option can be specified multiple times on the command line (arguments are concatenated).</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--skip-regions-file"></a><code>--skip-regions-file</code><a class="hash-link" href="#--skip-regions-file" title="Direct link to heading">#</a></h3><p>Option <code>--skip-regions-file</code> (short <code>-k</code>) specifies a file containing a list of genomic regions to skip (one per line). The file format can either be plain text (using the same input format as the <code>--regions</code> option), or BED format (i.e. tab-separated). GZipped files are also accepted.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--one-based-indexing"></a><code>--one-based-indexing</code><a class="hash-link" href="#--one-based-indexing" title="Direct link to heading">#</a></h3><p>Command <code>--one-based-indexing</code> directs the program to read all input regions (i.e. those specified in options <code>--regions</code>, <code>--regions-file</code>, <code>--skip-regions</code>, and <code>--skip-regions-file</code>) using 1-based indexing rather than 0-based indexing.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -T chr1:100 --one-based-indexing</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -T chr1:99 </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># equivalent to above</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--ignore-unmapped-contigs"></a><code>--ignore-unmapped-contigs</code><a class="hash-link" href="#--ignore-unmapped-contigs" title="Direct link to heading">#</a></h3><p>Command <code>--ignore-unmapped-contigs</code> can be used to force execution if there is a mismatch between the input reference genome and the one used to map input reads (as specified in the SAM header). In particular, if there are contigs present in the reference genome but not in the reference used for read mapping.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&lt;</span><span class="token plain">INFO</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> ------------------------------------------------------------------------</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&lt;</span><span class="token plain">INFO</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> octopus v0.6.1-beta </span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token plain">develop 9c97253c</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&lt;</span><span class="token plain">INFO</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> Copyright </span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token plain">c</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token plain"> </span><span class="token number" style="color:rgb(247, 140, 108)">2015</span><span class="token plain">-2019 University of Oxford</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&lt;</span><span class="token plain">INFO</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> ------------------------------------------------------------------------</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&lt;</span><span class="token plain">EROR</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> A user error has occurred:</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&lt;</span><span class="token plain">EROR</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&lt;</span><span class="token plain">EROR</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain">     Some or all of the contigs </span><span class="token keyword" style="font-style:italic">in</span><span class="token plain"> the reference genome </span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token plain">ref</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token plain"> are not</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&lt;</span><span class="token plain">EROR</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain">     present </span><span class="token keyword" style="font-style:italic">in</span><span class="token plain"> the </span><span class="token builtin class-name" style="color:rgb(255, 203, 107)">read</span><span class="token plain"> files.</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&lt;</span><span class="token plain">EROR</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&lt;</span><span class="token plain">EROR</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> To </span><span class="token builtin class-name" style="color:rgb(255, 203, 107)">help</span><span class="token plain"> resolve this error ensure the reference genome used </span><span class="token keyword" style="font-style:italic">for</span><span class="token plain"> mapping</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&lt;</span><span class="token plain">EROR</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> is the same as the one used </span><span class="token keyword" style="font-style:italic">for</span><span class="token plain"> calling </span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token plain">ref</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span><span class="token plain"> and all input contigs are</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&lt;</span><span class="token plain">EROR</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> present </span><span class="token keyword" style="font-style:italic">in</span><span class="token plain"> the </span><span class="token builtin class-name" style="color:rgb(255, 203, 107)">read</span><span class="token plain"> headers.</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain"></span><span class="token operator" style="color:rgb(137, 221, 255)">&lt;</span><span class="token plain">INFO</span><span class="token operator" style="color:rgb(137, 221, 255)">&gt;</span><span class="token plain"> ------------------------------------------------------------------------</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain" style="display:inline-block">
</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --ignore-unmapped-contigs </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># ignore error above</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain" style="display:inline-block">
</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--samples"></a><code>--samples</code><a class="hash-link" href="#--samples" title="Direct link to heading">#</a></h3><p>Option <code>--samples</code> (short <code>-S</code>) specifies a list of samples to use for variant calling. Each sample must be present in the input read files (using the <code>SM</code> SAM tag).</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -S sample1 sample2</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>If no samples are specified, all samples present in the input read files are used.</li><li>The option can be specified multiple times on the command line (arguments are concatenated).</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--samples-file"></a><code>--samples-file</code><a class="hash-link" href="#--samples-file" title="Direct link to heading">#</a></h3><p>Option <code>--samples-file</code> (short <code>-s</code>) specifies a file containing a list (one per line) of samples to use for variant calling.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -s samples.txt</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--output"></a><code>--output</code><a class="hash-link" href="#--output" title="Direct link to heading">#</a></h3><p>Option <code>--output</code> (short <code>-O</code>) sets the output destination. The file extension is used to determine the output file type (valid types are <code>.vcf</code>, <code>.vcf.gz</code>, and <code>.bcf</code>). If the file type is not recognised then uncompressed VCF is used.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -o calls.bcf</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>If no output is specified, <code>stdout</code> is used.</li><li>Like all path arguments in Octopus, the argument is assumed to be relative to the <code>--working-directory</code>, unless the path is absolute and exists.</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--contig-output-order"></a><code>--contig-output-order</code><a class="hash-link" href="#--contig-output-order" title="Direct link to heading">#</a></h3><p>Option <code>--contig-output-order</code> specifies the order that records will be processed and written to the output. Possible options are: <code>lexicographicalAscending</code>, <code>lexicographicalDescending</code>, <code>contigSizeAscending</code>, <code>contigSizeDescending</code>, <code>asInReferenceIndex</code>, <code>asInReferenceIndexReversed</code>, <code>unspecified</code></p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --contig-output-order asInReferenceIndexReversed</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--sites-only"></a><code>--sites-only</code><a class="hash-link" href="#--sites-only" title="Direct link to heading">#</a></h3><p>Command <code>--sites-only</code> removes genotype information from the final output (i.e. drops the VCF <code>FORMAT</code> and sample columns).</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --sites-only</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--bamout"></a><code>--bamout</code><a class="hash-link" href="#--bamout" title="Direct link to heading">#</a></h3><p>Option <code>--bamout</code> is used to produce <a href="/octopus/docs/guides/bamout">realigned evidence BAMs</a>. The option input is a file prefix where the BAMs should be written.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -o calls.vcf --bamout reads.realigned.bam </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># if &#x27;reads.bam&#x27; contains one sample</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I readsA.bam readsB.bam -o calls.vcf --bamout minibams</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>The <code>--bamout</code> command is only allowed if the final output is written to file (i.e. with <code>--output</code>).</li><li>The default behaviour of <code>--bamout</code> is to realign only those reads supporting variant haplotypes (see also <code>--full-bamout</code>).</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--bamout-type"></a><code>--bamout-type</code><a class="hash-link" href="#--bamout-type" title="Direct link to heading">#</a></h3><p>Option <code>--bamout-type</code> is used to select the type of realigned evidence BAM to produce when using the <code>--bamout</code> option. If <code>FULL</code> is chosen then all reads in the input BAM are written to the evidence BAM. If <code>MINI</code> is chosen (default) then only reads supporting variant sites are written.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -o calls.vcf --bamout reads.realigned.bam --bamout-type FULL</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--pedigree"></a><code>--pedigree</code><a class="hash-link" href="#--pedigree" title="Direct link to heading">#</a></h3><p>Option <code>--pedigree</code> is used to input a <a href="http://www.helsinki.fi/~tsjuntun/autogscan/pedigreefile.html" target="_blank" rel="noopener noreferrer">pedigree file</a> (<code>.ped</code>). Some calling models may use pedigree information to improve calling accuracy.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I mother.bam father.bam child.bam --pedigree family.ped </span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p>where <code>family.ped</code> might contain, for example</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI text"><div tabindex="0" class="prism-code language-text codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">my_trio FATHER  0   0   1   0</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">my_trio MOTHER  0   0   2   0</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">my_trio CHILD   FATHER  MOTHER  2   0</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>Specifying this option with a <code>.ped</code> file defining a trio relationship between three input samples will automatically activate the <code>trio</code> calling model.</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--fast"></a><code>--fast</code><a class="hash-link" href="#--fast" title="Direct link to heading">#</a></h3><p>Command <code>--fast</code> sets up Octopus for fast variant calling.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --fast</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Warning</strong> This command may reduce calling accuracy.</p><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--very-fast"></a><code>--very-fast</code><a class="hash-link" href="#--very-fast" title="Direct link to heading">#</a></h3><p>Command <code>--very-fast</code> sets up Octopus for very fast variant calling.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --very-fast</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain" style="display:inline-block">
</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Warning</strong> This command may reduce calling accuracy</p><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--data-profile"></a><code>--data-profile</code><a class="hash-link" href="#--data-profile" title="Direct link to heading">#</a></h3><p>Option <code>--data-profile</code> is used to generate a profile of the input read data, which may be used to make new <a href="/octopus/docs/guides/errorModels">sequence error models</a>.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -o calls.bcf --data-profile reads.profile.csv</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain" style="display:inline-block">
</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--bad-region-tolerance"></a><code>--bad-region-tolerance</code><a class="hash-link" href="#--bad-region-tolerance" title="Direct link to heading">#</a></h3><p>Option <code>--bad-region-tolerance</code> specifies the user tolerance for regions that may be &#x27;uncallable&#x27; (e.g. due to mapping errors) and slow down calling. The possible arguments are:</p><ul><li><code>LOW</code> Low tolerance to bad regions.</li><li><code>NORMAL</code> Default tolerance to bad regions.</li><li><code>HIGH</code> High tolerance to bad regions.</li><li><code>UNLIMITED</code> Turn off bad region detection.</li></ul><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --bad-region-tolerance UNLIMITED</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain" style="display:inline-block">
</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="read-pre-processing-options"></a>Read pre-processing options<a class="hash-link" href="#read-pre-processing-options" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--disable-read-preprocessing"></a><code>--disable-read-preprocessing</code><a class="hash-link" href="#--disable-read-preprocessing" title="Direct link to heading">#</a></h3><p>Command <code>--disable-read-preprocessing</code> can be used to disable all optional read preprocessing - all viable raw input alignments will be used.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --disable-read-preprocessing</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-base-quality"></a><code>--max-base-quality</code><a class="hash-link" href="#--max-base-quality" title="Direct link to heading">#</a></h3><p>Option <code>--max-base-quality</code> caps the base quality of all input reads.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-base-quality </span><span class="token number" style="color:rgb(247, 140, 108)">40</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--mask-low-quality-tails"></a><code>--mask-low-quality-tails</code><a class="hash-link" href="#--mask-low-quality-tails" title="Direct link to heading">#</a></h3><p>Option <code>--mask-low-quality-tails</code> is used to mask (assign base quality zero) read tail bases that have low base quality scores. The value provided to the option is the threshold used to define a &#x27;low&#x27; quality score.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --mask-low-quality-tails </span><span class="token number" style="color:rgb(247, 140, 108)">5</span><span class="token plain"> // use </span><span class="token number" style="color:rgb(247, 140, 108)">5</span><span class="token plain"> as minimum quality score</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --mask-low-quality-tails // use implicit </span><span class="token string" style="color:rgb(195, 232, 141)">&#x27;low&#x27;</span><span class="token plain"> score </span><span class="token punctuation" style="color:rgb(199, 146, 234)">(</span><span class="token plain">see -h</span><span class="token punctuation" style="color:rgb(199, 146, 234)">)</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--mask-tails"></a><code>--mask-tails</code><a class="hash-link" href="#--mask-tails" title="Direct link to heading">#</a></h3><p>Option <code>--mask-tails</code> is used to unconditionally mask (assign base quality zero) read tail bases. The value provided to the option is the number of bases to mask.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --mask-tails </span><span class="token number" style="color:rgb(247, 140, 108)">3</span><span class="token plain"> // mask the last </span><span class="token number" style="color:rgb(247, 140, 108)">3</span><span class="token plain"> bases of all reads</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--mask-soft-clipped-bases"></a><code>--mask-soft-clipped-bases</code><a class="hash-link" href="#--mask-soft-clipped-bases" title="Direct link to heading">#</a></h3><p>Command <code>--mask-soft-clipped-bases</code> is used to mask (assign base quality zero) all read bases that are soft clipped in the input alignments.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --mask-soft-clipped-bases</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--soft-clip-mask-threshold"></a><code>--soft-clip-mask-threshold</code><a class="hash-link" href="#--soft-clip-mask-threshold" title="Direct link to heading">#</a></h3><p>Option <code>--soft-clip-mask-threshold</code> makes the option <code>--soft-clip-masking</code> conditional on the base quality score; only bases below the value given to the option will be masked.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --soft-clip-masking </span><span class="token function" style="color:rgb(130, 170, 255)">yes</span><span class="token plain"> --soft-clip-mask-threshold </span><span class="token number" style="color:rgb(247, 140, 108)">10</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--mask-soft-clipped-boundary-bases"></a><code>--mask-soft-clipped-boundary-bases</code><a class="hash-link" href="#--mask-soft-clipped-boundary-bases" title="Direct link to heading">#</a></h3><p>Option <code>--mask-soft-clipped-boundary-bases</code> makes the option <code>--soft-clip-masking</code> mask addition bases adjacent to soft clipped bases. The value provided to the option is the number of additional bases to mask.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --soft-clip-masking </span><span class="token function" style="color:rgb(130, 170, 255)">yes</span><span class="token plain"> --mask-soft-clipped-boundary-bases </span><span class="token number" style="color:rgb(247, 140, 108)">5</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--mask-inverted-soft-clipping"></a><code>--mask-inverted-soft-clipping</code><a class="hash-link" href="#--mask-inverted-soft-clipping" title="Direct link to heading">#</a></h3><p>Command <code>--mask-inverted-soft-clipping</code> is used to mask (assign base quality zero) all read bases that are soft clipped <strong>and</strong> are inverted copies of nearby sequence.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --mask-inverted-soft-clipping</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--mask-3prime-shifted-soft-clipped-heads"></a><code>--mask-3prime-shifted-soft-clipped-heads</code><a class="hash-link" href="#--mask-3prime-shifted-soft-clipped-heads" title="Direct link to heading">#</a></h3><p>Command <code>--mask-3prime-shifted-soft-clipped-heads</code> is used to mask (assign base quality zero) all read bases that are soft clipped <strong>and</strong> are shifted copies of nearby 3&#x27; sequence.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --mask-3prime-shifted-soft-clipped-heads</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--disable-adapter-masking"></a><code>--disable-adapter-masking</code><a class="hash-link" href="#--disable-adapter-masking" title="Direct link to heading">#</a></h3><p>Command <code>--disable-adapter-masking</code> is used to mask (assign base quality zero) read bases that are considered to be adapter contamination.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --disable-adapter-masking</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>The algorithm used to detect adapter contamination depends only on the input alignment mapping information; no library of adapter sequences are used.</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--disable-overlap-masking"></a><code>--disable-overlap-masking</code><a class="hash-link" href="#--disable-overlap-masking" title="Direct link to heading">#</a></h3><p>Command <code>--disable-overlap-masking</code> is used to mask (assign base quality zero) read bases of read templates that contain overlapping segments, in order to remove non-independent base observations. </p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --disable-overlap-masking</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>All but of the overlapping read bases are masked, leaving one base untouched. If two segments are overlapping, then half of the 5&#x27; bases of each segment are masked.</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--split-long-reads"></a><code>--split-long-reads</code><a class="hash-link" href="#--split-long-reads" title="Direct link to heading">#</a></h3><p>Command <code>--split-long-reads</code> stipulates that reads longer than <a href="#option---max-read-length"><code>--max-read-length</code></a> should be split into smaller linked reads.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-read-length </span><span class="token number" style="color:rgb(247, 140, 108)">200</span><span class="token plain"> --split-long-reads</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--consider-unmapped-reads"></a><code>--consider-unmapped-reads</code><a class="hash-link" href="#--consider-unmapped-reads" title="Direct link to heading">#</a></h3><p>Command <code>--consider-unmapped-reads</code> turns off the read filter that removes reads marked as unmapped in the input alignments.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --consider-unmapped-reads</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-mapping-quality"></a><code>--min-mapping-quality</code><a class="hash-link" href="#--min-mapping-quality" title="Direct link to heading">#</a></h3><p>Option <code>--min-mapping-quality</code> specifies the minimum mapping quality that reads must have to be considered; reads with mapping quality below this will be filtered and not used for analysis.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --min-mapping-quality </span><span class="token number" style="color:rgb(247, 140, 108)">10</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--good-base-quality"></a><code>--good-base-quality</code><a class="hash-link" href="#--good-base-quality" title="Direct link to heading">#</a></h3><p>Option <code>--good-base-quality</code> defines the minimum quality of a &#x27;good&#x27; base for the options <code>--min-good-base-fraction</code> and <code>--min-good-bases</code>.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --good-base-quality </span><span class="token number" style="color:rgb(247, 140, 108)">10</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-good-base-fraction"></a><code>--min-good-base-fraction</code><a class="hash-link" href="#--min-good-base-fraction" title="Direct link to heading">#</a></h3><p>Option <code>--min-good-base-fraction</code> specifies the fraction of &#x27;good&#x27; (see <code>--good-base-quality</code>) base qualities a read must have in order to be considered; reads with a fraction of &#x27;good&#x27; base qualities less than this will be filtered and not considered.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --min-good-base-fraction </span><span class="token number" style="color:rgb(247, 140, 108)">0.3</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-good-bases"></a><code>--min-good-bases</code><a class="hash-link" href="#--min-good-bases" title="Direct link to heading">#</a></h3><p>Option <code>--min-good-bases</code> specifies the number of &#x27;good&#x27; (see <code>--good-base-quality</code>) base qualities a read must have in order to be considered; reads with &#x27;good&#x27; base qualities less than this will be filtered and not considered.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --min-good-bases </span><span class="token number" style="color:rgb(247, 140, 108)">50</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--allow-qc-fails"></a><code>--allow-qc-fails</code><a class="hash-link" href="#--allow-qc-fails" title="Direct link to heading">#</a></h3><p>Command <code>--allow-qc-fails</code> permits reads marked as QC fail in the input alignments to be considered, otherwise they will be filtered.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --allow-qc-fails</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-read-length"></a><code>--min-read-length</code><a class="hash-link" href="#--min-read-length" title="Direct link to heading">#</a></h3><p>Option <code>--min-read-length</code> specifies the minimum length (number of sequence bases) of reads to be considered; reads with length less than this will be filtered and not considered.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --min-read-length </span><span class="token number" style="color:rgb(247, 140, 108)">50</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-read-length"></a><code>--max-read-length</code><a class="hash-link" href="#--max-read-length" title="Direct link to heading">#</a></h3><p>Option <code>--max-read-length</code> specifies the maximum length (number of sequence bases) of reads to be considered; reads with length greater than this will be filtered and not considered.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-read-length </span><span class="token number" style="color:rgb(247, 140, 108)">500</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--allow-marked-duplicates"></a><code>--allow-marked-duplicates</code><a class="hash-link" href="#--allow-marked-duplicates" title="Direct link to heading">#</a></h3><p>Command <code>--allow-marked-duplicates</code> disables the filter that removes reads marked as duplicate in the input alignments. The default behaviour is to remove reads marked duplicate.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --allow-marked-duplicates</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--allow-octopus-duplicates"></a><code>--allow-octopus-duplicates</code><a class="hash-link" href="#--allow-octopus-duplicates" title="Direct link to heading">#</a></h3><p>Command <code>--allow-octopus-duplicates</code> disables the filter that removes reads that Octopus considers to be duplicates. The default behaviour is to remove duplicate reads.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --allow-octopus-duplicates</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--duplicate-read-detection-policy"></a><code>--duplicate-read-detection-policy</code><a class="hash-link" href="#--duplicate-read-detection-policy" title="Direct link to heading">#</a></h3><p>Option <code>--duplicate-read-detection-policy</code> specifies approach to use for detecting <a href="/octopus/docs/guides/preprocessing">duplicate reads</a>. Possible arguments are:</p><ul><li><code>RELAXED</code> Require 5&#x27; mapping co-ordinate matches <strong>and</strong> identical cigar strings.</li><li><code>AGGRESSIVE</code> Require 5&#x27; mapping co-ordinate matches only.</li></ul><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --duplicate-read-detection-policy AGGRESSIVE</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--allow-secondary-alignments"></a><code>--allow-secondary-alignments</code><a class="hash-link" href="#--allow-secondary-alignments" title="Direct link to heading">#</a></h3><p>Command <code>--allow-secondary-alignments</code> disables the filter that removes reads that are marked as secondary alignments in the input alignments. The default behaviour is to remove reads marked secondary.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --allow-secondary-alignments</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--allow-supplementary-alignments"></a><code>--allow-supplementary-alignments</code><a class="hash-link" href="#--allow-supplementary-alignments" title="Direct link to heading">#</a></h3><p>Command <code>--allow-supplementary-alignments</code> disables the filter that removes reads that are marked as supplementary alignments in the input alignments. The default behaviour is to remove reads marked supplementary.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --allow-supplementary-alignments</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-decoy-supplementary-alignment-mapping-quality"></a><code>--max-decoy-supplementary-alignment-mapping-quality</code><a class="hash-link" href="#--max-decoy-supplementary-alignment-mapping-quality" title="Direct link to heading">#</a></h3><p>Option <code>--max-decoy-supplementary-alignment-mapping-quality</code> removes any reads with supplementary alignments (i.e. <code>SA</code> SAM tag) to decoy contigs with mapping quality greater than the specified value.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-decoy-supplementary-alignment-mapping-quality </span><span class="token number" style="color:rgb(247, 140, 108)">20</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-unplaced-supplementary-alignment-mapping-quality"></a><code>--max-unplaced-supplementary-alignment-mapping-quality</code><a class="hash-link" href="#--max-unplaced-supplementary-alignment-mapping-quality" title="Direct link to heading">#</a></h3><p>Command <code>--max-unplaced-supplementary-alignment-mapping-quality removes reads with supplementary alignments (i.e. </code>SA` SAM tag) to unplaced contigs with mapping quality greater than the specified value.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-unplaced-supplementary-alignment-mapping-quality </span><span class="token number" style="color:rgb(247, 140, 108)">20</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-unlocalized-supplementary-alignment-mapping-quality"></a><code>--max-unlocalized-supplementary-alignment-mapping-quality</code><a class="hash-link" href="#--max-unlocalized-supplementary-alignment-mapping-quality" title="Direct link to heading">#</a></h3><p>Command <code>--max-unlocalized-supplementary-alignment-mapping-quality</code> removes reads with supplementary alignments (i.e. <code>SA</code> SAM tag) to unlocalized contigs with mapping quality greater than the specified value.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-unlocalized-supplementary-alignment-mapping-quality </span><span class="token number" style="color:rgb(247, 140, 108)">20</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--no-reads-with-unmapped-segments"></a><code>--no-reads-with-unmapped-segments</code><a class="hash-link" href="#--no-reads-with-unmapped-segments" title="Direct link to heading">#</a></h3><p>Command <code>--no-reads-with-unmapped-segments</code> removes reads that are marked as having unmapped segments in the input alignments. The default behaviour is to allow reads with unmapped segments.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --no-reads-with-unmapped-segments</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--no-reads-with-distant-segments"></a><code>--no-reads-with-distant-segments</code><a class="hash-link" href="#--no-reads-with-distant-segments" title="Direct link to heading">#</a></h3><p>Command <code>--no-reads-with-distant-segments</code> removes reads that have segments mapped to different contigs. The default behaviour is to allow reads with distant segments.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --no-reads-with-distant-segments</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--no-adapter-contaminated-reads"></a><code>--no-adapter-contaminated-reads</code><a class="hash-link" href="#--no-adapter-contaminated-reads" title="Direct link to heading">#</a></h3><p>Command <code>--no-adapter-contaminated-reads</code> removes reads that are considered to have adapter contamination (i.e. sequence bases from the adapter). The default behaviour is to allow reads with adapter contamination.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --no-adapter-contaminated-reads</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--disable-downsampling"></a><code>--disable-downsampling</code><a class="hash-link" href="#--disable-downsampling" title="Direct link to heading">#</a></h3><p>Command <code>--disable-downsampling</code> turns off downsampling.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --disable-downsampling</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--downsample-above"></a><code>--downsample-above</code><a class="hash-link" href="#--downsample-above" title="Direct link to heading">#</a></h3><p>Option <code>--downsample-above</code> specifies the read depth required to mark a position as a candidate for the downsampler.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --downsample-above </span><span class="token number" style="color:rgb(247, 140, 108)">5000</span><span class="token plain"> </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># depths up to 5000x are permitted</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--downsample-target"></a><code>--downsample-target</code><a class="hash-link" href="#--downsample-target" title="Direct link to heading">#</a></h3><p>Option <code>--downsample-target</code> specifies the target read depth for the downsampler for all candidate sites. Reads will be removed from the input alignments until all positions have read depth not greater than this.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --downsample-target </span><span class="token number" style="color:rgb(247, 140, 108)">100</span><span class="token plain"> </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># downsample to 100x</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--use-same-read-profile-for-all-samples"></a><code>--use-same-read-profile-for-all-samples</code><a class="hash-link" href="#--use-same-read-profile-for-all-samples" title="Direct link to heading">#</a></h3><p>Command <code>--use-same-read-profile-for-all-samples</code> specifies that the same input read profile should be used for all samples, rather than generating one for each sample. This essentially means that the same read distribution is assumed for all samples. </p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --use-same-read-profile-for-all-samples</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="variant-discovery-options"></a>Variant discovery options<a class="hash-link" href="#variant-discovery-options" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--variant-discovery-mode"></a><code>--variant-discovery-mode</code><a class="hash-link" href="#--variant-discovery-mode" title="Direct link to heading">#</a></h3><p>Option <code>--variant-discovery-mode</code> specifies the thresholds used for candidate variant discovery, which affects the overall sensitivity of the generators. Possible values are:</p><ul><li><code>ILLUMINA</code> The default mode, for Illumina quality reads.</li><li><code>PACBIO</code> For PacBio quality reads - requires more observations to propose a candidate, particular indel candidates.</li></ul><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --variant-discovery-mode PACBIO</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--disable-denovo-variant-discovery"></a><code>--disable-denovo-variant-discovery</code><a class="hash-link" href="#--disable-denovo-variant-discovery" title="Direct link to heading">#</a></h3><p>Command <code>--disable-denovo-variant-discovery</code> disables the pileup candidate variant generator.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --disable-denovo-variant-discovery</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--disable-repeat-candidate-generator"></a><code>--disable-repeat-candidate-generator</code><a class="hash-link" href="#--disable-repeat-candidate-generator" title="Direct link to heading">#</a></h3><p>Command <code>--disable-repeat-candidate-generator</code> disables the tandem repeat candidate variant generator.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --disable-repeat-candidate-generator</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--disable-assembly-candidate-generator"></a><code>--disable-assembly-candidate-generator</code><a class="hash-link" href="#--disable-assembly-candidate-generator" title="Direct link to heading">#</a></h3><p>Command <code>--disable-assembly-candidate-generator</code> disables the local <em>de novo</em> assembly candidate variant generator.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --disable-assembly-candidate-generator</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--source-candidates"></a><code>--source-candidates</code><a class="hash-link" href="#--source-candidates" title="Direct link to heading">#</a></h3><p>Option <code>--source-candidates</code> (short <code>-c</code>) accepts a list of VCF files, the contents of which will be added to the candidate variant set.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --source-candidates calls.vcf.gz</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>The option accepts files in the <code>.vcf</code>, <code>.vcf.gz</code>, and <code>.bcf</code> formats. Files in the <code>.vcf.gz</code> and <code>.bcf</code> format must be indexed. For larger</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--source-candidates-file"></a><code>--source-candidates-file</code><a class="hash-link" href="#--source-candidates-file" title="Direct link to heading">#</a></h3><p>Option <code>--source-candidates-file</code> can be used to provide one or more files containing lists (one per line) of files in the <code>.vcf</code>, <code>.vcf.gz</code>, and <code>.bcf</code> formats (see notes on <code>--source-candidates</code>).</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --source-candidates-file vcfs.txt</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p>where <code>vcfs.txt</code> may contain, for example</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI text"><div tabindex="0" class="prism-code language-text codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">/path/to/vcfs/callsA.vcf.gz</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">/path/to/vcfs/callsB.bcf</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-source-candidate-quality"></a><code>--min-source-candidate-quality</code><a class="hash-link" href="#--min-source-candidate-quality" title="Direct link to heading">#</a></h3><p>Option <code>--min-source-candidate-quality</code> specifies the minimum <code>QUAL</code> score for a user provided source variant (see <code>--source-candidates</code> and <code>--source-candidates-file</code>) to be added to the final candidate variant list; any records with <code>QUAL</code> less than this will not be considered.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -c calls.vcf.gz --min-source-candidate-quality </span><span class="token number" style="color:rgb(247, 140, 108)">10</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--use-filtered-source-candidates"></a><code>--use-filtered-source-candidates</code><a class="hash-link" href="#--use-filtered-source-candidates" title="Direct link to heading">#</a></h3><p>Command <code>--use-filtered-source-candidates</code> specifies allows variants in the user-provided candidate variants (see <code>--source-candidates</code> and <code>--source-candidates-file</code>) that are marked as filtered (according to the <code>FILTER</code> column) should be added to the final candidate variant list. The default behaviour is to remove filtered variants.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -c calls.vcf.gz --use-filtered-source-candidates</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-pileup-base-quality"></a><code>--min-pileup-base-quality</code><a class="hash-link" href="#--min-pileup-base-quality" title="Direct link to heading">#</a></h3><p>Option <code>--min-pileup-base-quality</code> specifies the minimum base quality that a SNV in the input alignments must have in ordered to be considered by the pileup candidate variant generator.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --min-pileup-base-quality </span><span class="token number" style="color:rgb(247, 140, 108)">10</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-supporting-reads"></a><code>--min-supporting-reads</code><a class="hash-link" href="#--min-supporting-reads" title="Direct link to heading">#</a></h3><p>Option <code>--min-supporting-reads</code> specifies the minimum number of supporting reads a variant must have in the input alignments to be included in the candidate variant list from the pileup candidate variant generator.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --min-supporting-reads </span><span class="token number" style="color:rgb(247, 140, 108)">3</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--allow-pileup-candidates-from-likely-misaligned-reads"></a><code>--allow-pileup-candidates-from-likely-misaligned-reads</code><a class="hash-link" href="#--allow-pileup-candidates-from-likely-misaligned-reads" title="Direct link to heading">#</a></h3><p>Command <code>--allow-pileup-candidates-from-likely-misaligned-reads</code> stops the pileup candidate generator filtering out candidates that are considered likely to originate from read misalignment, which can otherwise result in many false candidates.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --allow-pileup-candidates-from-likely-misaligned-reads</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-variant-size"></a><code>--max-variant-size</code><a class="hash-link" href="#--max-variant-size" title="Direct link to heading">#</a></h3><p>Option <code>--max-variant-size</code> specifies the maximum size (w.r.t reference region) a candidate variant can have to be considered by the calling algorithm; candidate variants with size greater than this will be removed and not considered.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-variant-size </span><span class="token number" style="color:rgb(247, 140, 108)">500</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--kmer-sizes"></a><code>--kmer-sizes</code><a class="hash-link" href="#--kmer-sizes" title="Direct link to heading">#</a></h3><p>Option <code>--kmer-sizes</code> specifies the default kmer sizes to try for local <em>de novo</em> assembly. Assembly graphs will be constructed for each kmer size in all active assembler regions, and the union of candidate variants from each assembler graph used for the final candidate set.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --kmer-sizes </span><span class="token number" style="color:rgb(247, 140, 108)">5</span><span class="token plain"> </span><span class="token number" style="color:rgb(247, 140, 108)">10</span><span class="token plain"> </span><span class="token number" style="color:rgb(247, 140, 108)">15</span><span class="token plain"> </span><span class="token number" style="color:rgb(247, 140, 108)">20</span><span class="token plain"> </span><span class="token number" style="color:rgb(247, 140, 108)">25</span><span class="token plain"> </span><span class="token number" style="color:rgb(247, 140, 108)">30</span><span class="token plain"> </span><span class="token number" style="color:rgb(247, 140, 108)">35</span><span class="token plain"> </span><span class="token number" style="color:rgb(247, 140, 108)">40</span><span class="token plain"> </span><span class="token number" style="color:rgb(247, 140, 108)">45</span><span class="token plain"> </span><span class="token number" style="color:rgb(247, 140, 108)">50</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--num-fallback-kmers"></a><code>--num-fallback-kmers</code><a class="hash-link" href="#--num-fallback-kmers" title="Direct link to heading">#</a></h3><p>Option <code>--num-fallback-kmers</code> specifies the number of additional kmer sizes to try in the case that none of the default (i.e. kmer sizes specified in <code>--kmer-sizes</code>) is able to construct a valid assembly graph (often the case for smaller kmer sizes).</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --num-fallback-kmers </span><span class="token number" style="color:rgb(247, 140, 108)">20</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--fallback-kmer-gap"></a><code>--fallback-kmer-gap</code><a class="hash-link" href="#--fallback-kmer-gap" title="Direct link to heading">#</a></h3><p>Option <code>--fallback-kmer-gap</code> speifies the increment between fallback kmer sizes (see <code>--num-fallback-kmers</code>)</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --fallback-kmer-gap </span><span class="token number" style="color:rgb(247, 140, 108)">5</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --kmer-sizes </span><span class="token number" style="color:rgb(247, 140, 108)">5</span><span class="token plain"> </span><span class="token number" style="color:rgb(247, 140, 108)">10</span><span class="token plain"> --num-fallback-kmers </span><span class="token number" style="color:rgb(247, 140, 108)">2</span><span class="token plain"> --fallback-kmer-gap </span><span class="token number" style="color:rgb(247, 140, 108)">5</span><span class="token plain"> </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># always try kmer sizes 5 and 10, and then try 15 and 20 if 5 and 10 don&#x27;t work</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-region-to-assemble"></a><code>--max-region-to-assemble</code><a class="hash-link" href="#--max-region-to-assemble" title="Direct link to heading">#</a></h3><p>Option <code>--max-region-to-assemble</code> specifies the maximum reference region size that can be used to construct a local <em>de novo</em> assembly graph. Larger values enable detection of larger variants (e.g. large deletions), but increase the likelihood that small kmer sizes will result in an invalid assembly graph, and therefore decrease sensitivity for smaller variation.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-region-to-assemble </span><span class="token number" style="color:rgb(247, 140, 108)">1000</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-assemble-region-overlap"></a><code>--max-assemble-region-overlap</code><a class="hash-link" href="#--max-assemble-region-overlap" title="Direct link to heading">#</a></h3><p>Option <code>--max-assemble-region-overlap</code> specifies the maximum overlap between reference regions used to build local <em>de novo</em> assembly graphs. Larger overlaps result in more assembly graphs being constructed and may increase sensitivity for variation, at the expense of compute time. </p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-assemble-region-overlap </span><span class="token number" style="color:rgb(247, 140, 108)">100</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--assemble-all"></a><code>--assemble-all</code><a class="hash-link" href="#--assemble-all" title="Direct link to heading">#</a></h3><p>Command <code>--assemble-all</code> forces all reference regions to be used for local <em>de novo</em> assembly, rather than only regions considered likely to contain variation.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --assemble-all</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Warning</strong> This command may result in substantially longer runtimes. </p><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--assembler-mask-base-quality"></a><code>--assembler-mask-base-quality</code><a class="hash-link" href="#--assembler-mask-base-quality" title="Direct link to heading">#</a></h3><p>Option <code>--assembler-mask-base-quality</code> specifies the minimum base quality an aligned base should have to avoid being &#x27;masked&#x27; (i.e. converted to reference) before being inserted into the local <em>de novo</em> assembly graph. Higher values reduce sensitivity to noise and increase the likelihood of finding bubbles in graphs with larger kmer sizes, as the cost of decreased sensitivity at lower kmer sizes. </p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --assembler-mask-base-quality </span><span class="token number" style="color:rgb(247, 140, 108)">20</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--allow-cycles"></a><code>--allow-cycles</code><a class="hash-link" href="#--allow-cycles" title="Direct link to heading">#</a></h3><p>Command <code>--allow-cycles</code> forces the assembler to consider assembly graphs containing non-reference cycles, which usually result in false candidates.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --allow-cycles</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-kmer-prune"></a><code>--min-kmer-prune</code><a class="hash-link" href="#--min-kmer-prune" title="Direct link to heading">#</a></h3><p>Option <code>--min-kmer-prune</code> specifies the minimum number of kmer observations that must be present in the local <em>de novo</em> assembly graph for the kmer to stay in the graph before variant extraction. Lower values increase sensitivity but lower specificity.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --min-kmer-prune </span><span class="token number" style="color:rgb(247, 140, 108)">3</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-bubbles"></a><code>--max-bubbles</code><a class="hash-link" href="#--max-bubbles" title="Direct link to heading">#</a></h3><p>Option <code>--max-bubbles</code> specifies the maximum number of bubbles in the final local <em>de novo</em> assembly graph that may be explored for candidate variant generation. Higher values increase sensitivity but reduce specificity.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-bubbles </span><span class="token number" style="color:rgb(247, 140, 108)">100</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-bubble-score"></a><code>--min-bubble-score</code><a class="hash-link" href="#--min-bubble-score" title="Direct link to heading">#</a></h3><p>Option <code>--min-bubble-score</code> specifies the minimum score a bubble explored in a local <em>de novo</em> assembly graph must have for the bubble to be extracted as a candidate variant. The &#x27;bubble score&#x27; is the average of all kmer observations along the bubble edge, scaled by the probability the kmer observations along the bubble edge originate from a single read strand. The bubble score can be viewed as a proxy for the number of &#x27;good&#x27; read observations for the bubble (i.e. variant), and therefore higher values increase specificity but reduce sensitivity.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --min-bubble-score </span><span class="token number" style="color:rgb(247, 140, 108)">5</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-candidate-credible-vaf-probability"></a><code>--min-candidate-credible-vaf-probability</code><a class="hash-link" href="#--min-candidate-credible-vaf-probability" title="Direct link to heading">#</a></h3><p>Option <code>--min-candidate-credible-vaf-probability</code> sets the minimum probability mass above <code>--min-credible-somatic-frequency</code> required to &#x27;discover&#x27; a variant when using the <code>cancer</code> calling model. Smaller values increase the number of candidate variants generated, potentially improving sensitivity but also increasing computational complexity.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -N NORMAL --min-candidate-credible-vaf-probability </span><span class="token number" style="color:rgb(247, 140, 108)">0.5</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="haplotype-generation-options"></a>Haplotype generation options<a class="hash-link" href="#haplotype-generation-options" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-haplotypes"></a><code>--max-haplotypes</code><a class="hash-link" href="#--max-haplotypes" title="Direct link to heading">#</a></h3><p>Option <code>--max-haplotypes</code> specifies the maximum number of haplotypes that are considered by the calling model. It also specifies the target number of haplotypes that the haplotype generator should produce on each iteration of the algorithm. If the haplotype generator is unable to satisfy the request (i.e. produces a greater number of haplotypes), then the set of haplotypes is reduced to this size by removing haplotypes considered unlikely using several likelihood based statistics.</p><p>Increasing <code>--max-haplotypes</code> reduces the chance that a true haplotype is incorrectly filtered before evaluation by the calling model. It also increases the number of candidate variants that may be considered on each iteration of the algorithm, potentially improving calling accuracy. However, increasing this value will usually increase runtimes - sometimes substantially. </p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-haplotypes </span><span class="token number" style="color:rgb(247, 140, 108)">500</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--haplotype-holdout-threshold"></a><code>--haplotype-holdout-threshold</code><a class="hash-link" href="#--haplotype-holdout-threshold" title="Direct link to heading">#</a></h3><p>Option <code>--haplotype-holdout-threshold</code> specifies the number of haplotypes that the haplotype generator can produce before some active candidate variants are added to the holdout stack. The value must not be less than <code>--max-haplotypes</code>.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --haplotype-holdout-threshold </span><span class="token number" style="color:rgb(247, 140, 108)">1000</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--haplotype-overflow"></a><code>--haplotype-overflow</code><a class="hash-link" href="#--haplotype-overflow" title="Direct link to heading">#</a></h3><p>Option <code>--haplotype-overflow</code> specifies the number of haplotypes that the haplotype generator can produce before the current active region must be skipped. The value must not be less than <code>--haplotype-holdout-threshold</code>.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --haplotype-overflow </span><span class="token number" style="color:rgb(247, 140, 108)">5000</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-holdout-depth"></a><code>--max-holdout-depth</code><a class="hash-link" href="#--max-holdout-depth" title="Direct link to heading">#</a></h3><p>Option <code>--max-holdout-depth</code> specifies the maximum size of the holdout stack.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-holdout-depth </span><span class="token number" style="color:rgb(247, 140, 108)">0</span><span class="token plain"> </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># no holdouts</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--extension-level"></a><code>--extension-level</code><a class="hash-link" href="#--extension-level" title="Direct link to heading">#</a></h3><p>Option <code>--extension-level</code> specifies the condition for extending the active haplotype tree with novel alleles. The possible values are:</p><ul><li><code>MINIMAL</code> Only include novel alleles overlapping with overlapping reads. </li><li><code>NORMAL</code> Include novel alleles with reads overlapping the rightmost included allele.</li><li><code>AGGRESSIVE</code> No conditions on extension other than the number of alleles.</li></ul><p>More aggressive extension levels result in larger haplotype blocks, which may improve phase lengths and accuracy. However, aggressive extension increases the possibility of including novel alleles that cannot be phased with active alleles which will increase compute time.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --extension-level AGGRESSIVE</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--lagging-level"></a><code>--lagging-level</code><a class="hash-link" href="#--lagging-level" title="Direct link to heading">#</a></h3><p>Option <code>--lagging-level</code> specifies the extent to which active alleles remain active in the next algorithm iteration, increasing the length of haplotype blocks.</p><ul><li><code>NONE</code> Disable lagging; each iteration of the algorithm will evaluate a novel set of alleles.</li><li><code>NORMAL</code> Consider previous active alleles if there are reads overlapping them with the next active allele set.</li><li><code>AGGRESSIVE</code> Consider previous active alleles if there are overlapping reads that span them, and the next active alleles.</li></ul><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --lagging-level AGGRESSIVE</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--backtrack-level"></a><code>--backtrack-level</code><a class="hash-link" href="#--backtrack-level" title="Direct link to heading">#</a></h3><p>Option <code>--backtrack-level</code> specifies the extent to which </p><ul><li><code>NONE</code> Disable all backtracking.</li><li><code>NORMAL</code> Enables backtracking.</li><li><code>AGGRESSIVE</code> Currently the same as <code>NORMAL</code>.</li></ul><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --backtrack-level NORMAL</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-protected-haplotype-posterior"></a><code>--min-protected-haplotype-posterior</code><a class="hash-link" href="#--min-protected-haplotype-posterior" title="Direct link to heading">#</a></h3><p>Option <code>--min-protected-haplotype-posterior</code> specifies the minimum posterior probability that a haplotype is present in the samples (according to the calling model) for the haplotype to avoid being removed from consideration; haplotypes with posterior probability less than this may be filtered. Increasing the value of this option results in a greater number of haplotypes being filtered, allowing the haplotype tree to grow to include more candidate alleles, and reducing computational complexity. However, larger values also increase the chance that a true haplotype is incorrectly discarded. </p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --min-protected-haplotype-posterior 1e-5</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--dont-protect-reference-haplotype"></a><code>--dont-protect-reference-haplotype</code><a class="hash-link" href="#--dont-protect-reference-haplotype" title="Direct link to heading">#</a></h3><p>Command <code>--dont-protect-reference-haplotype</code> disables protection of the reference haplotype during haplotype filtering, and therefore ensures that the reference haplotype is always considered by the calling model.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --dont-protect-reference-haplotype</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--bad-region-tolerance-1"></a><code>--bad-region-tolerance</code><a class="hash-link" href="#--bad-region-tolerance-1" title="Direct link to heading">#</a></h3><p>Option <code>--bad-region-tolerance</code> specifies the &#x27;tolerance&#x27; for spending time calling variants in regions that are unlikely to be callable (e.g. due to low complexity sequence). Such regions tend to be computationally difficult and skipping them can save a lot of wasted computation time. However, identification of such regions is not perfect and could result in skipping regions with real variation. Possible argument values are:</p><ul><li><code>LOW</code> Skip region that show any signs of being uncallable.</li><li><code>NORMAL</code> Skip region that show reasonable signs of being uncallable.</li><li><code>HIGH</code> Skip region that show strong signs of being uncallable.</li></ul><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --bad-region-tolerance LOW</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="common-variant-calling-options"></a>Common variant calling options<a class="hash-link" href="#common-variant-calling-options" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--caller"></a><code>--caller</code><a class="hash-link" href="#--caller" title="Direct link to heading">#</a></h3><p>Option <code>--caller</code> (short &#x27;-C&#x27;) specifies the calling model to be used. The option must only be set if the calling model is not automatically determined from other options.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --caller cancer </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># e.g. for tumour-only</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--organism-ploidy"></a><code>--organism-ploidy</code><a class="hash-link" href="#--organism-ploidy" title="Direct link to heading">#</a></h3><p>Option <code>--organism-ploidy</code> (short &#x27;-P&#x27;) specifies the default ploidy of all input samples. All contigs will be assumed to have this ploidy unless specified otherwise in <code>--contig-ploidies</code>.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --organism-ploidy </span><span class="token number" style="color:rgb(247, 140, 108)">3</span><span class="token plain"> </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># triploid calling</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--contig-ploidies"></a><code>--contig-ploidies</code><a class="hash-link" href="#--contig-ploidies" title="Direct link to heading">#</a></h3><p>Option <code>--contig-ploidies</code> (short <code>-p</code>) can be used to specify the ploidy of contigs (format <code>contig=ploidy</code>) that are not the same as the <code>--organism-ploidy</code>, or the ploidy of individual sample contigs (format <code>sample:contig=ploidy</code>).</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --contig-ploidies </span><span class="token assign-left variable" style="color:rgb(191, 199, 213)">X</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain"> </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># Contig &quot;X&quot; has ploidy 1 for all samples</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --contig-ploidies HG002:X</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain"> </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># Contig &quot;X&quot; has ploidy 1 for sample &quot;HG002&quot;</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--contig-ploidies-file"></a><code>--contig-ploidies-file</code><a class="hash-link" href="#--contig-ploidies-file" title="Direct link to heading">#</a></h3><p>Option <code>--contig-ploidies-file</code> can be used to provide a file specifying contig or sample contig ploidies (see <code>--contig-ploidies</code>), one per line.  </p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --contig-ploidies-file ploidies.txt</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p>where <code>ploidies.txt</code> could contain</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI text"><div tabindex="0" class="prism-code language-text codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">X=1</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">HG002:X=1</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-variant-posterior"></a><code>--min-variant-posterior</code><a class="hash-link" href="#--min-variant-posterior" title="Direct link to heading">#</a></h3><p>Option <code>--min-variant-posterior</code> specifies the minimum posterior probability (Phred scale) required to call a variant. Candidate variants with posterior probability less than this will not be reported in the final call set.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --min-variant-posterior </span><span class="token number" style="color:rgb(247, 140, 108)">0.5</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>For calling models with more than one class of variation, this option refers to the calling models default variant class. For example, in the <code>cancer</code> calling model there are both germline and somatic variants, and this option refers to germline variants (see the option <code>--min-somatic-posterior</code> for somatic variants).</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--refcall"></a><code>--refcall</code><a class="hash-link" href="#--refcall" title="Direct link to heading">#</a></h3><p>Option <code>--refcall</code> can be used to enable reference calling, meaning Octopus will generate a &#x27;gVCF&#x27; format output. There are two possible arguments for this option:</p><ul><li><code>BLOCKED</code> Merge adjacent called reference positions with similar quality (see <code>--refcall-block-merge-quality</code>) a single gVCF record.</li><li><code>POSITIONAL</code> Emit a gVCF record for all positions.</li></ul><p>If the option is specified but no argument is provide, <code>BLOCKED</code> is assumed.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --refcall BLOCKED</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --refcall </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># same as above</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --refcall POSITIONAL</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--refcall-block-merge-quality"></a><code>--refcall-block-merge-quality</code><a class="hash-link" href="#--refcall-block-merge-quality" title="Direct link to heading">#</a></h3><p>Option <code>--refcall-block-merge-quality</code> specifies the quality (Phred scale) threshold for merging adjacent called reference positions when <code>BLOCKED</code> refcalls are requested; adjacent reference positions with an absolute quality difference less or equal than this will be merged into a block.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --refcall --refcall-block-merge-quality </span><span class="token number" style="color:rgb(247, 140, 108)">20</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-refcall-posterior"></a><code>--min-refcall-posterior</code><a class="hash-link" href="#--min-refcall-posterior" title="Direct link to heading">#</a></h3><p>Option <code>--min-refcall-posterior</code> specifies the minimum posterior probability required to call a position as homozygous reference; positions with posterior probability less than this will not be reported in the output gVCF.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --min-refcall-posterior </span><span class="token number" style="color:rgb(247, 140, 108)">10</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-refcall-posterior"></a><code>--max-refcall-posterior</code><a class="hash-link" href="#--max-refcall-posterior" title="Direct link to heading">#</a></h3><p>Option <code>--max-refcall-posterior</code> caps the <code>QUAL</code> of all reference calls, which may result in larger reference blocks and smaller gVCF file sizes.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-refcall-posterior </span><span class="token number" style="color:rgb(247, 140, 108)">100</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--snp-heterozygosity"></a><code>--snp-heterozygosity</code><a class="hash-link" href="#--snp-heterozygosity" title="Direct link to heading">#</a></h3><p>Option <code>--snp-heterozygosity</code> specifies the SNV heterozygosity parameter for the Coalesent mutation model, used to assign prior probabilities.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --snp-heterozygosity </span><span class="token number" style="color:rgb(247, 140, 108)">0.1</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--snp-heterozygosity-stdev"></a><code>--snp-heterozygosity-stdev</code><a class="hash-link" href="#--snp-heterozygosity-stdev" title="Direct link to heading">#</a></h3><p>Option <code>--snp-heterozygosity-stdev</code> specifies the standard deviation of the SNV heterozygosity parameter (see <code>--snp-heterozygosity</code>).</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --snp-heterozygosity-stdev </span><span class="token number" style="color:rgb(247, 140, 108)">0.1</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--indel-heterozygosity"></a><code>--indel-heterozygosity</code><a class="hash-link" href="#--indel-heterozygosity" title="Direct link to heading">#</a></h3><p>Option <code>--indel-heterozygosity</code> specifies the INDEL heterozygosity parameter for the Coalesent mutation model, used to assign prior probabilities.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --indel-heterozygosity </span><span class="token number" style="color:rgb(247, 140, 108)">0.001</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-genotypes"></a><code>--max-genotypes</code><a class="hash-link" href="#--max-genotypes" title="Direct link to heading">#</a></h3><p>Option <code>--max-genotypes</code> specifies the maximum number of candidate genotypes that must be evaluated by the calling model. If there are more possible candidate genotypes than this value then the algorithm may decide to remove some candidate genotypes using heuristics. The number of candidate genotypes to evaluate can have a substantial impact on runtime for some calling models (e.g. <code>cancer</code>).</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-genotypes </span><span class="token number" style="color:rgb(247, 140, 108)">1000</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-genotype-combinations"></a><code>--max-genotype-combinations</code><a class="hash-link" href="#--max-genotype-combinations" title="Direct link to heading">#</a></h3><p>Option <code>--max-genotype-combinations</code> specifies the maximum number of candidate joint genotype vectors that must be evaluated by the calling model. If there are more possible candidate joint genotype vectors than this value the algorithm may decide to remove some candidate genotype vectors using heuristics. The number of candidate joint genotype vectors to evaluate can have a substantial impact on runtime. </p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-genotype-combinations </span><span class="token number" style="color:rgb(247, 140, 108)">100000</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>This option is only used by calling models that consider joint genotypes (i.e. <code>population</code> and <code>trio</code>)</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--use-uniform-genotype-priors"></a><code>--use-uniform-genotype-priors</code><a class="hash-link" href="#--use-uniform-genotype-priors" title="Direct link to heading">#</a></h3><p>Command <code>--use-uniform-genotype-priors</code> indicates that the uniform genotype prior model should be used for calling genotypes and variants.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --use-uniform-genotype-priors</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--use-independent-genotype-priors"></a><code>--use-independent-genotype-priors</code><a class="hash-link" href="#--use-independent-genotype-priors" title="Direct link to heading">#</a></h3><p>Command <code>--use-independent-genotype-priors</code> indicates that an independent genotype prior model should be used for evaluating joint genotype vectors.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --use-independent-genotype-priors</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--model-posterior"></a><code>--model-posterior</code><a class="hash-link" href="#--model-posterior" title="Direct link to heading">#</a></h3><p>Option <code>--model-posterior</code> enables or disables model posterior evaluation at each called variant site.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --model-posterior </span><span class="token function" style="color:rgb(130, 170, 255)">yes</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--disable-inactive-flank-scoring"></a><code>--disable-inactive-flank-scoring</code><a class="hash-link" href="#--disable-inactive-flank-scoring" title="Direct link to heading">#</a></h3><p>Command <code>--disable-inactive-flank-scoring</code> disables an additional step during haplotype likelihood calculation that attempts to correct low likelihoods caused by inactive variation in the flanking regions of the haplotype under evaluation.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --disable-inactive-flank-scoring</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--dont-model-mapping-quality"></a><code>--dont-model-mapping-quality</code><a class="hash-link" href="#--dont-model-mapping-quality" title="Direct link to heading">#</a></h3><p>Command <code>--dont-model-mapping-quality</code> disables consideration of mapping quality in the haplotype likelihood calculation. This can improve calling accuracy if read mapping qualities are well calibrated.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --dont-model-mapping-quality</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--sequence-error-model"></a><code>--sequence-error-model</code><a class="hash-link" href="#--sequence-error-model" title="Direct link to heading">#</a></h3><p>Option <code>--sequence-error-model</code> specifies the <a href="/octopus/docs/guides/errorModels">sequence error model</a> to use.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --sequence-error-model PCR.NOVASEQ </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># built-in error model</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --sequence-error-model /path/to/my/error.model </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># custom error model</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>The same error model is used for all input reads. </li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-indel-errors"></a><code>--max-indel-errors</code><a class="hash-link" href="#--max-indel-errors" title="Direct link to heading">#</a></h3><p>Option <code>--max-indel-errors</code> specifies the maximum number of indel errors in an individual read fragment that can be accuretely modelled by the haplotype likelihood model. Larger values usually require greater computational resources (determined by your systems available SIMD instructions).</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-indel-errors </span><span class="token number" style="color:rgb(247, 140, 108)">32</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--use-wide-hmm-scores"></a><code>--use-wide-hmm-scores</code><a class="hash-link" href="#--use-wide-hmm-scores" title="Direct link to heading">#</a></h3><p>Command <code>--use-wide-hmm-scores</code> sets the score variable computed by the pair HMM for haplotype likelihoods to 32 bits (the default is 16 bits). This can avoid score overflow in long noisy reads, but will slow does the computation.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --use-wide-hmm-scores</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-vb-seeds"></a><code>--max-vb-seeds</code><a class="hash-link" href="#--max-vb-seeds" title="Direct link to heading">#</a></h3><p>Option <code>--max-vb-seeds</code> specifies the maximum number of seeds that Variational Bayes models can use for posterior evaluation. Increasing the number of seeds increases the likelihood that a posterior mode will be identified, but results in more computation time.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-vb-seeds </span><span class="token number" style="color:rgb(247, 140, 108)">50</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--read-linkage"></a><code>--read-linkage</code><a class="hash-link" href="#--read-linkage" title="Direct link to heading">#</a></h3><p>Option <code>--read-linkage</code> specifies how reads are linked in the input alignments. Read linkage information is used by the haplotype likelihood calculation and can improve calling accuracy and increase phase lengths.</p><ul><li><code>NONE</code> Reads are not linked in any way.</li><li><code>PAIRED</code> Reads may be paired, with pairs having identical read names.</li><li><code>LINKED</code> Reads may be linked or paired, with linked reads having identical <code>BX</code> tags.</li></ul><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --read-linkage LINKED</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-phase-score"></a><code>--min-phase-score</code><a class="hash-link" href="#--min-phase-score" title="Direct link to heading">#</a></h3><p>Option <code>--min-phase-score</code> specifies the minimum phase score (<code>PQ</code> in VCF) required to emit adjacent variant calls in the same <a href="/octopus/docs/guides/advanced/vcf#Haplotypes">phase set</a>. Increasing this value results in less sites being phased, but reduces the phase false positive rate.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --min-phase-score </span><span class="token number" style="color:rgb(247, 140, 108)">5</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--disable-early-phase-detection"></a><code>--disable-early-phase-detection</code><a class="hash-link" href="#--disable-early-phase-detection" title="Direct link to heading">#</a></h3><p>Command <code>--disable-early-phase-detection</code> prevents the phasing algorithm being applied to partially resolved haplotype blocks, which can lead to removal of complete phased segments from the head of the current haplotype block. This heuristic can prevent discontiguous phase blocks being resolved, which are more likely in some data (e.g. linked reads).</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --disable-early-phase-detection</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="cancer-variant-calling-options"></a>Cancer variant calling options<a class="hash-link" href="#cancer-variant-calling-options" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--normal-samples"></a><code>--normal-samples</code><a class="hash-link" href="#--normal-samples" title="Direct link to heading">#</a></h3><p>Option <code>--normal-samples</code> specifies which of the input samples are normal samples for tumour-normal paired analysis.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -N NORMAL</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>Specifying this option will automatically activate the <code>cancer</code> calling model.</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-somatic-haplotypes"></a><code>--max-somatic-haplotypes</code><a class="hash-link" href="#--max-somatic-haplotypes" title="Direct link to heading">#</a></h3><p>Option <code>--max-somatic-haplotypes</code> specifies the maximum number of unique haplotypes containing somatic variation that can be modelled by the somatic genotype model. If there are more true somatic haplotypes present in the input data than this value then the model will not accuretely fit the data and true somatic variants may not be called, however, larger values substantially increase the computational complexity of the model and potentially increase the false positive rate.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --max-somatic-haplotypes </span><span class="token number" style="color:rgb(247, 140, 108)">3</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--somatic-snv-prior"></a><code>--somatic-snv-prior</code><a class="hash-link" href="#--somatic-snv-prior" title="Direct link to heading">#</a></h3><p>Option <code>--somatic-snv-prior</code> specifies the somatic SNV mutation prior probability for the samples under consideration.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --somatic-snv-prior 1e-7</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--somatic-indel-prior"></a><code>--somatic-indel-prior</code><a class="hash-link" href="#--somatic-indel-prior" title="Direct link to heading">#</a></h3><p>Option <code>--somatic-indel-prior</code> specifies the somatic INDEL mutation prior probability for the samples under consideration.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --somatic-indel-prior 1e-7</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-expected-somatic-frequency"></a><code>--min-expected-somatic-frequency</code><a class="hash-link" href="#--min-expected-somatic-frequency" title="Direct link to heading">#</a></h3><p>Option <code>--min-expected-somatic-frequency</code> specifies the minimum expected Variant Allele Frequency (VAF) for somatic mutations in the samples. This value is used by the <code>cancer</code> calling model as the lower bound for the VAF posterior marginalisation used to compute the posterior probability that a candidate variant is a somatic mutation. Decreasing this value increases sensitivity for somatic mutations with smaller VAFs, but also increases sensitivity to noise.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -N NORMAL --min-expected-somatic-frequency </span><span class="token number" style="color:rgb(247, 140, 108)">0.05</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-credible-somatic-frequency"></a><code>--min-credible-somatic-frequency</code><a class="hash-link" href="#--min-credible-somatic-frequency" title="Direct link to heading">#</a></h3><p>Option <code>--min-credible-somatic-frequency</code> specifies the minimum credible Variant Allele Frequency (VAF) for somatic mutations in the samples. This value is used for candidate variant discovery, and also by the <code>cancer</code> calling model as the lower-bound on candidate somatic mutation VAF credible regions (i.e. if the credible region computed for a mutation contains VAFs less than this value then the mutation will not be called as somatic). Decreasing this option increases sensitivity for somatic mutations with lower VAFs, but also increases sensitivity to noise and increases computational complexity as more candidate variants will be generated.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -N NORMAL --min-credible-somatic-frequency </span><span class="token number" style="color:rgb(247, 140, 108)">0.001</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>If the value provided to this option is greater than the value specified by <code>--min-expected-somatic-frequency</code>, then this value is used for both options.</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--tumour-germline-concentration"></a><code>--tumour-germline-concentration</code><a class="hash-link" href="#--tumour-germline-concentration" title="Direct link to heading">#</a></h3><p>Option <code>--tumour-germline-concentration</code> sets the Dirichlet concentration parameter for the germline haplotypes of tumour samples. Larger values concentrate more prior probability mass on equal frequencies of germline haplotypes and also increase prior mass on low VAFs for somatic haplotypes. This can help the model correctly classify somatic variation when the normal sample is not informative (or not present), but also reduces sensitivity to somatic variation with larger VAFs.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -N NORMAL --tumour-germline-concentration </span><span class="token number" style="color:rgb(247, 140, 108)">5</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--somatic-credible-mass"></a><code>--somatic-credible-mass</code><a class="hash-link" href="#--somatic-credible-mass" title="Direct link to heading">#</a></h3><p>Option <code>--somatic-credible-mass</code> specifies the probability mass to used to compute the credible interval for determining whether to call a variant as somatic (see also <code>--min-credible-somatic-frequency</code>). Larger values result in wider credible regions, increasing sensitivity for lower VAF somatic mutations, but also increase sensitivity to noise. </p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -N NORMAL --somatic-credible-mass </span><span class="token number" style="color:rgb(247, 140, 108)">0.99</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-somatic-posterior"></a><code>--min-somatic-posterior</code><a class="hash-link" href="#--min-somatic-posterior" title="Direct link to heading">#</a></h3><p>Option <code>--min-somatic-posterior</code> specifies the minimum posterior probability (Phred scale) required to call a candidate variant as a somatic mutation.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -N NORMAL --min-somatic-posterior </span><span class="token number" style="color:rgb(247, 140, 108)">1</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--normal-contamination-risk"></a><code>--normal-contamination-risk</code><a class="hash-link" href="#--normal-contamination-risk" title="Direct link to heading">#</a></h3><p>Option <code>--normal-contamination-risk</code> indicates the risk that the normal sample contains contamination from the tumour samples. There are two possible values:</p><ul><li><code>LOW</code> The algorithm will not consider normal contamination when generating candidate genotypes.</li><li><code>HIGH</code> The algorithm will consider normal contamination when generating candidate genotypes.</li></ul><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -N NORMAL --normal-contamination-risk HIGH</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--somatics-only"></a><code>--somatics-only</code><a class="hash-link" href="#--somatics-only" title="Direct link to heading">#</a></h3><p>Command <code>--somatics-only</code> indicates that only variant sites called as somatic mutations (i.e. tagged <code>SOMATIC</code>) should appear in the final output.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -N NORMAL --somatics-only</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Warning</strong> Using this command will produce a VCF file that cannot be re-filtered using Octopus.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="trio-variant-calling-options"></a>Trio variant calling options<a class="hash-link" href="#trio-variant-calling-options" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--maternal-sample"></a><code>--maternal-sample</code><a class="hash-link" href="#--maternal-sample" title="Direct link to heading">#</a></h3><p>Option <code>--maternal-sample</code> (short &#x27;-M&#x27;) indicates which of the input samples is the mother of the proband. If this option is specified then <code>--paternal-sample</code> must also be specified.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -M mother -F father</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>Specifying this option will automatically activate the <code>trio</code> calling model.</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--paternal-sample"></a><code>--paternal-sample</code><a class="hash-link" href="#--paternal-sample" title="Direct link to heading">#</a></h3><p>Option <code>--paternal-sample</code> (short &#x27;-F&#x27;) indicates which of the input samples is the father of the proband. If this option is specified then <code>--maternal-sample</code> must also be specified.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -M mother -F father</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Notes</strong></p><ul><li>Specifying this option will automatically activate the <code>trio</code> calling model.</li></ul><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--denovo-snv-prior"></a><code>--denovo-snv-prior</code><a class="hash-link" href="#--denovo-snv-prior" title="Direct link to heading">#</a></h3><p>Option <code>--denovo-snv-prior</code> specifies the <em>de novo</em> SNV mutation prior probability for the samples under consideration.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --ped trio.ped --denovo-snv-prior 1e-7</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--denovo-indel-prior"></a><code>--denovo-indel-prior</code><a class="hash-link" href="#--denovo-indel-prior" title="Direct link to heading">#</a></h3><p>Option <code>--denovo-indel-prior</code> specifies the <em>de novo</em> INDEL mutation prior probability for the samples under consideration.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --ped trio.ped --denovo-indel-prior 1e-8</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-denovo-posterior"></a><code>--min-denovo-posterior</code><a class="hash-link" href="#--min-denovo-posterior" title="Direct link to heading">#</a></h3><p>Option <code>--min-denovo-posterior</code> specifies the minimum posterior probability (Phred scale) required to call a candidate variant as a <em>de novo</em> mutation.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --ped trio.ped --min-denovo-posterior </span><span class="token number" style="color:rgb(247, 140, 108)">2</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--denovos-only"></a><code>--denovos-only</code><a class="hash-link" href="#--denovos-only" title="Direct link to heading">#</a></h3><p>Command <code>--denovos-only</code> indicates that only variant sites called as <em>de novo</em> mutations (i.e. tagged <code>DENOVO</code>) should appear in the final output.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --ped trio.ped --denovos-only</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Warning</strong> Using this command will produce a VCF file that cannot be re-filtered using Octopus.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="polyclone-variant-calling-options"></a>Polyclone variant calling options<a class="hash-link" href="#polyclone-variant-calling-options" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-clones"></a><code>--max-clones</code><a class="hash-link" href="#--max-clones" title="Direct link to heading">#</a></h3><p>Option <code>--max-clones</code> specifies the maximum number of clones that can be modelled by the <code>polyclone</code> calling model. If there are more clones than this present in the input data then the model will not fit the data well and some true variation may not be called. However, larger values increase the computational complexity of the model and also increase sensitivity to noise.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --caller polyclone --max-clones </span><span class="token number" style="color:rgb(247, 140, 108)">5</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-clone-frequency"></a><code>--min-clone-frequency</code><a class="hash-link" href="#--min-clone-frequency" title="Direct link to heading">#</a></h3><p>Option <code>--min-clone-frequency</code> specifies the lower-bound Variant Allele Frequency (VAF) to use when computing the posterior probability for a variant. Smaller values increase sensitivity and the false positive rate.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --caller polyclone --min-clone-frequency </span><span class="token number" style="color:rgb(247, 140, 108)">0.05</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--clone-prior"></a><code>--clone-prior</code><a class="hash-link" href="#--clone-prior" title="Direct link to heading">#</a></h3><p>Option <code>--clone-prior</code> sets the prior probability for each new clone proposed by the model</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --caller polyclone --clone-prior </span><span class="token number" style="color:rgb(247, 140, 108)">0.1</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--clone-concentration"></a><code>--clone-concentration</code><a class="hash-link" href="#--clone-concentration" title="Direct link to heading">#</a></h3><p>Option <code>--clone-concentration </code> sets the concentration parameter for the symmetric Dirichlet distribution used to model clone frequencies</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --caller polyclone --clone-concentration </span><span class="token number" style="color:rgb(247, 140, 108)">10</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="cell-variant-calling-options"></a>Cell variant calling options<a class="hash-link" href="#cell-variant-calling-options" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-copy-loss"></a><code>--max-copy-loss</code><a class="hash-link" href="#--max-copy-loss" title="Direct link to heading">#</a></h3><p>Option <code>--max-copy-loss</code> specifies the maximum number of germline haplotypes losses that can be considered by the model.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --caller cell --max-copy-loss </span><span class="token number" style="color:rgb(247, 140, 108)">1</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--max-copy-gains"></a><code>--max-copy-gains</code><a class="hash-link" href="#--max-copy-gains" title="Direct link to heading">#</a></h3><p>Option <code>--max-copy-gains </code> specifies the maximum number of haplotypes gains that can be considered by the model.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --caller cell --max-copy-gains </span><span class="token number" style="color:rgb(247, 140, 108)">1</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--somatic-cnv-prior"></a><code>--somatic-cnv-prior</code><a class="hash-link" href="#--somatic-cnv-prior" title="Direct link to heading">#</a></h3><p>Option <code>--somatic-cnv-prior </code> sets the prior probability of a loci having a copy-number change.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --caller cell --somatic-cnv-prior 1e-10</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--dropout-concentration"></a><code>--dropout-concentration</code><a class="hash-link" href="#--dropout-concentration" title="Direct link to heading">#</a></h3><p>Option <code>--dropout-concentration</code> sets the default Dirichlet concentration prior on haplotype frequencies. The higher the concentration parameter, the more probability mass around the centre of the distribution (0.5 for diploid). Lowering the concentration parameter means thereâ€™s more mass on extreme frequencies, which in turn means the model is less sensitive to dropout, but also less sensitive to real somatic variation. Setting to unity would put a uniform prior on frequencies.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --caller cell --dropout-concentration </span><span class="token number" style="color:rgb(247, 140, 108)">10</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--sample-dropout-concentration"></a><code>--sample-dropout-concentration</code><a class="hash-link" href="#--sample-dropout-concentration" title="Direct link to heading">#</a></h3><p>Option <code>--sample-dropout-concentration</code> sets the Dirichlet concentration prior on haplotype frequencies for a specific sample, otherwise <code>--dropout-concentration</code> is used.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --caller cell --sample-dropout-concentration </span><span class="token assign-left variable" style="color:rgb(191, 199, 213)">NORMAL</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token number" style="color:rgb(247, 140, 108)">100</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--phylogeny-concentration"></a><code>--phylogeny-concentration</code><a class="hash-link" href="#--phylogeny-concentration" title="Direct link to heading">#</a></h3><p>Option <code>--phylogeny-concentration</code> sets the symmetric Dirichlet concentration prior on group mixture proportions in each phylogeny. A larger concentration parameter implies a more even distribution of samples across the tree.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --caller cell --phylogeny-concentration </span><span class="token number" style="color:rgb(247, 140, 108)">10</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="variant-filtering-options"></a>Variant filtering options<a class="hash-link" href="#variant-filtering-options" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--disable-call-filtering"></a><code>--disable-call-filtering</code><a class="hash-link" href="#--disable-call-filtering" title="Direct link to heading">#</a></h3><p>Command <code>--disable-call-filtering</code> disables variant call filtering.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --disable-call-filtering</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--filter-expression"></a><code>--filter-expression</code><a class="hash-link" href="#--filter-expression" title="Direct link to heading">#</a></h3><p>Option <code>--filter-expression</code> sets the threshold filter expression to use for filtering variants not tagged with <code>SOMATIC</code> or <code>DENOVO</code>.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --filter-expression </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;QUAL &lt; 10&quot;</span><span class="token plain"> </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># PASS calls with QUAL &gt;= 10</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--somatic-filter-expression"></a><code>--somatic-filter-expression</code><a class="hash-link" href="#--somatic-filter-expression" title="Direct link to heading">#</a></h3><p>Option <code>--somatic-filter-expression</code> sets the threshold filter expression to use for filtering <code>SOMATIC</code> variants.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -N NORMAL --somatic-filter-expression </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;QUAL &lt; 20 | PP &lt; 20&quot;</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--denovo-filter-expression"></a><code>--denovo-filter-expression</code><a class="hash-link" href="#--denovo-filter-expression" title="Direct link to heading">#</a></h3><p>Option <code>--denovo-filter-expression</code> sets the threshold filter expression to use for filtering <code>DENOVO</code> variants.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --ped trio.ped --denovo-filter-expression </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;QUAL &lt; 20 | PP &lt; 20&quot;</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--refcall-filter-expression"></a><code>--refcall-filter-expression</code><a class="hash-link" href="#--refcall-filter-expression" title="Direct link to heading">#</a></h3><p>Option <code>--refcall-filter-expression</code> sets the threshold filter expression to use for filtering sites called homozygous reference.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --refcall --refcall-filter-expression </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;QUAL &lt; 10&quot;</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--use-preprocessed-reads-for-filtering"></a><code>--use-preprocessed-reads-for-filtering</code><a class="hash-link" href="#--use-preprocessed-reads-for-filtering" title="Direct link to heading">#</a></h3><p>Command <code>--use-preprocessed-reads-for-filtering</code> forces use of the same read pre-processing steps used for calling variants for filtering variants; otherwise all well-formed reads are used for filtering.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --use-preprocessed-reads-for-filtering</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--keep-unfiltered-calls"></a><code>--keep-unfiltered-calls</code><a class="hash-link" href="#--keep-unfiltered-calls" title="Direct link to heading">#</a></h3><p>Command <code>--keep-unfiltered-calls</code> requests that Octopus keep a copy of the VCF file produced before filtering is applied. The copy has <code>unfiltered</code> appended to the final output name.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -o calls.vcf --keep-unfiltered-calls </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># writes calls.unfiltered.vcf</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--annotations"></a><code>--annotations</code><a class="hash-link" href="#--annotations" title="Direct link to heading">#</a></h3><p>Option <code>--annotations</code> requests that the values of a sub-set of the measures used for filtering are reported in the final VCF output.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --annotations SB </span><span class="token comment" style="color:rgb(105, 112, 152);font-style:italic"># adds SB FORMAT field to each record</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--filter-vcf"></a><code>--filter-vcf</code><a class="hash-link" href="#--filter-vcf" title="Direct link to heading">#</a></h3><p>Option <code>--filter-vcf</code> specifies an Octopus VCF file to filter. No calling is performed.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --filter-vcf calls.bcf</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--forest-model"></a><code>--forest-model</code><a class="hash-link" href="#--forest-model" title="Direct link to heading">#</a></h3><p>Option <code>--forest-model</code> enables <a href="/octopus/docs/guides/filtering/forest">random forest variant filtering</a>. The argument to the option is a ranger forest file.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    --forest-model octopus/resources/forests/germline.v0.6.3-beta.forest</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--somatic-forest-model"></a><code>--somatic-forest-model</code><a class="hash-link" href="#--somatic-forest-model" title="Direct link to heading">#</a></h3><p>Option <code>--somatic-forest-model</code> enables <a href="/octopus/docs/guides/filtering/forest">random forest variant filtering</a> for somatic variants. The argument to the option is a ranger forest file.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam -N NORMAL </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    --forest-model octopus/resources/forests/germline.v0.6.3-beta.forest</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    --somatic-forest-model octopus/resources/forests/somatic.v0.6.3-beta.forest</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--min-forest-quality"></a><code>--min-forest-quality</code><a class="hash-link" href="#--min-forest-quality" title="Direct link to heading">#</a></h3><p>Option <code>--min-forest-quality</code> specifies the random forest minimum quality score (phred scale) required to PASS a variant call (<code>RFGQ_ALL</code>), and each samples genotype calls (<code>RFGQ</code>).</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I reads.bam --min-forest-quality </span><span class="token number" style="color:rgb(247, 140, 108)">7</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h3><a aria-hidden="true" tabindex="-1" class="anchor" id="--use-germline-forest-for-somatic-normals"></a><code>--use-germline-forest-for-somatic-normals</code><a class="hash-link" href="#--use-germline-forest-for-somatic-normals" title="Direct link to heading">#</a></h3><p>Command <code>--use-germline-forest-for-somatic-normals</code> specifies that the forest model given to <code>--forest-model</code> should be used to score normal sample genotypes for in somatic records, rather than the forest model given to <code>--somatic-forest-model</code>.</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R ref.fa -I normal.bam tumour.bam -N NORMAL --forest germline.forest --somatic-forest somatic.forest -o calls.vcf --use-germline-forest-for-somatic-normals</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div></div></article><div class="margin-vert--xl"><div class="row"><div class="col"><a href="https://github.com/${organizationName}/${projectName}/edit/${branch}/website/docs/cli.md" target="_blank" rel="noreferrer noopener"><svg fill="currentColor" height="1.2em" width="1.2em" preserveAspectRatio="xMidYMid meet" role="img" viewBox="0 0 40 40" class="iconEdit_mS5F" aria-label="Edit page"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div></div></div><div class="margin-vert--lg"><nav class="pagination-nav" aria-label="Docs pages navigation"><div class="pagination-nav__item"></div><div class="pagination-nav__item pagination-nav__item--next"></div></nav></div></div></div><div class="col col--3"><div class="tableOfContents_vrFS thin-scrollbar"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#general-options" class="table-of-contents__link">General options</a><ul><li><a href="#--help" class="table-of-contents__link"><code>--help</code></a></li><li><a href="#--version" class="table-of-contents__link"><code>--version</code></a></li><li><a href="#--config" class="table-of-contents__link"><code>--config</code></a></li><li><a href="#--trace" class="table-of-contents__link"><code>--trace</code></a></li><li><a href="#--working-directory" class="table-of-contents__link"><code>--working-directory</code></a></li><li><a href="#--resolve-symlinks" class="table-of-contents__link"><code>--resolve-symlinks</code></a></li><li><a href="#--threads" class="table-of-contents__link"><code>--threads</code></a></li><li><a href="#--max-reference-cache-memory" class="table-of-contents__link"><code>--max-reference-cache-memory</code></a></li><li><a href="#--target-read-buffer-memory" class="table-of-contents__link"><code>--target-read-buffer-memory</code></a></li><li><a href="#--target-working-memory" class="table-of-contents__link"><code>--target-working-memory</code></a></li><li><a href="#--temp-directory-prefix" class="table-of-contents__link"><code>--temp-directory-prefix</code></a></li><li><a href="#--reference" class="table-of-contents__link"><code>--reference</code></a></li><li><a href="#--reads" class="table-of-contents__link"><code>--reads</code></a></li><li><a href="#--reads-file" class="table-of-contents__link"><code>--reads-file</code></a></li><li><a href="#--regions" class="table-of-contents__link"><code>--regions</code></a></li><li><a href="#--regions-file" class="table-of-contents__link"><code>--regions-file</code></a></li><li><a href="#--skip-regions" class="table-of-contents__link"><code>--skip-regions</code></a></li><li><a href="#--skip-regions-file" class="table-of-contents__link"><code>--skip-regions-file</code></a></li><li><a href="#--one-based-indexing" class="table-of-contents__link"><code>--one-based-indexing</code></a></li><li><a href="#--ignore-unmapped-contigs" class="table-of-contents__link"><code>--ignore-unmapped-contigs</code></a></li><li><a href="#--samples" class="table-of-contents__link"><code>--samples</code></a></li><li><a href="#--samples-file" class="table-of-contents__link"><code>--samples-file</code></a></li><li><a href="#--output" class="table-of-contents__link"><code>--output</code></a></li><li><a href="#--contig-output-order" class="table-of-contents__link"><code>--contig-output-order</code></a></li><li><a href="#--sites-only" class="table-of-contents__link"><code>--sites-only</code></a></li><li><a href="#--bamout" class="table-of-contents__link"><code>--bamout</code></a></li><li><a href="#--bamout-type" class="table-of-contents__link"><code>--bamout-type</code></a></li><li><a href="#--pedigree" class="table-of-contents__link"><code>--pedigree</code></a></li><li><a href="#--fast" class="table-of-contents__link"><code>--fast</code></a></li><li><a href="#--very-fast" class="table-of-contents__link"><code>--very-fast</code></a></li><li><a href="#--data-profile" class="table-of-contents__link"><code>--data-profile</code></a></li><li><a href="#--bad-region-tolerance" class="table-of-contents__link"><code>--bad-region-tolerance</code></a></li></ul></li><li><a href="#read-pre-processing-options" class="table-of-contents__link">Read pre-processing options</a><ul><li><a href="#--disable-read-preprocessing" class="table-of-contents__link"><code>--disable-read-preprocessing</code></a></li><li><a href="#--max-base-quality" class="table-of-contents__link"><code>--max-base-quality</code></a></li><li><a href="#--mask-low-quality-tails" class="table-of-contents__link"><code>--mask-low-quality-tails</code></a></li><li><a href="#--mask-tails" class="table-of-contents__link"><code>--mask-tails</code></a></li><li><a href="#--mask-soft-clipped-bases" class="table-of-contents__link"><code>--mask-soft-clipped-bases</code></a></li><li><a href="#--soft-clip-mask-threshold" class="table-of-contents__link"><code>--soft-clip-mask-threshold</code></a></li><li><a href="#--mask-soft-clipped-boundary-bases" class="table-of-contents__link"><code>--mask-soft-clipped-boundary-bases</code></a></li><li><a href="#--mask-inverted-soft-clipping" class="table-of-contents__link"><code>--mask-inverted-soft-clipping</code></a></li><li><a href="#--mask-3prime-shifted-soft-clipped-heads" class="table-of-contents__link"><code>--mask-3prime-shifted-soft-clipped-heads</code></a></li><li><a href="#--disable-adapter-masking" class="table-of-contents__link"><code>--disable-adapter-masking</code></a></li><li><a href="#--disable-overlap-masking" class="table-of-contents__link"><code>--disable-overlap-masking</code></a></li><li><a href="#--split-long-reads" class="table-of-contents__link"><code>--split-long-reads</code></a></li><li><a href="#--consider-unmapped-reads" class="table-of-contents__link"><code>--consider-unmapped-reads</code></a></li><li><a href="#--min-mapping-quality" class="table-of-contents__link"><code>--min-mapping-quality</code></a></li><li><a href="#--good-base-quality" class="table-of-contents__link"><code>--good-base-quality</code></a></li><li><a href="#--min-good-base-fraction" class="table-of-contents__link"><code>--min-good-base-fraction</code></a></li><li><a href="#--min-good-bases" class="table-of-contents__link"><code>--min-good-bases</code></a></li><li><a href="#--allow-qc-fails" class="table-of-contents__link"><code>--allow-qc-fails</code></a></li><li><a href="#--min-read-length" class="table-of-contents__link"><code>--min-read-length</code></a></li><li><a href="#--max-read-length" class="table-of-contents__link"><code>--max-read-length</code></a></li><li><a href="#--allow-marked-duplicates" class="table-of-contents__link"><code>--allow-marked-duplicates</code></a></li><li><a href="#--allow-octopus-duplicates" class="table-of-contents__link"><code>--allow-octopus-duplicates</code></a></li><li><a href="#--duplicate-read-detection-policy" class="table-of-contents__link"><code>--duplicate-read-detection-policy</code></a></li><li><a href="#--allow-secondary-alignments" class="table-of-contents__link"><code>--allow-secondary-alignments</code></a></li><li><a href="#--allow-supplementary-alignments" class="table-of-contents__link"><code>--allow-supplementary-alignments</code></a></li><li><a href="#--max-decoy-supplementary-alignment-mapping-quality" class="table-of-contents__link"><code>--max-decoy-supplementary-alignment-mapping-quality</code></a></li><li><a href="#--max-unplaced-supplementary-alignment-mapping-quality" class="table-of-contents__link"><code>--max-unplaced-supplementary-alignment-mapping-quality</code></a></li><li><a href="#--max-unlocalized-supplementary-alignment-mapping-quality" class="table-of-contents__link"><code>--max-unlocalized-supplementary-alignment-mapping-quality</code></a></li><li><a href="#--no-reads-with-unmapped-segments" class="table-of-contents__link"><code>--no-reads-with-unmapped-segments</code></a></li><li><a href="#--no-reads-with-distant-segments" class="table-of-contents__link"><code>--no-reads-with-distant-segments</code></a></li><li><a href="#--no-adapter-contaminated-reads" class="table-of-contents__link"><code>--no-adapter-contaminated-reads</code></a></li><li><a href="#--disable-downsampling" class="table-of-contents__link"><code>--disable-downsampling</code></a></li><li><a href="#--downsample-above" class="table-of-contents__link"><code>--downsample-above</code></a></li><li><a href="#--downsample-target" class="table-of-contents__link"><code>--downsample-target</code></a></li><li><a href="#--use-same-read-profile-for-all-samples" class="table-of-contents__link"><code>--use-same-read-profile-for-all-samples</code></a></li></ul></li><li><a href="#variant-discovery-options" class="table-of-contents__link">Variant discovery options</a><ul><li><a href="#--variant-discovery-mode" class="table-of-contents__link"><code>--variant-discovery-mode</code></a></li><li><a href="#--disable-denovo-variant-discovery" class="table-of-contents__link"><code>--disable-denovo-variant-discovery</code></a></li><li><a href="#--disable-repeat-candidate-generator" class="table-of-contents__link"><code>--disable-repeat-candidate-generator</code></a></li><li><a href="#--disable-assembly-candidate-generator" class="table-of-contents__link"><code>--disable-assembly-candidate-generator</code></a></li><li><a href="#--source-candidates" class="table-of-contents__link"><code>--source-candidates</code></a></li><li><a href="#--source-candidates-file" class="table-of-contents__link"><code>--source-candidates-file</code></a></li><li><a href="#--min-source-candidate-quality" class="table-of-contents__link"><code>--min-source-candidate-quality</code></a></li><li><a href="#--use-filtered-source-candidates" class="table-of-contents__link"><code>--use-filtered-source-candidates</code></a></li><li><a href="#--min-pileup-base-quality" class="table-of-contents__link"><code>--min-pileup-base-quality</code></a></li><li><a href="#--min-supporting-reads" class="table-of-contents__link"><code>--min-supporting-reads</code></a></li><li><a href="#--allow-pileup-candidates-from-likely-misaligned-reads" class="table-of-contents__link"><code>--allow-pileup-candidates-from-likely-misaligned-reads</code></a></li><li><a href="#--max-variant-size" class="table-of-contents__link"><code>--max-variant-size</code></a></li><li><a href="#--kmer-sizes" class="table-of-contents__link"><code>--kmer-sizes</code></a></li><li><a href="#--num-fallback-kmers" class="table-of-contents__link"><code>--num-fallback-kmers</code></a></li><li><a href="#--fallback-kmer-gap" class="table-of-contents__link"><code>--fallback-kmer-gap</code></a></li><li><a href="#--max-region-to-assemble" class="table-of-contents__link"><code>--max-region-to-assemble</code></a></li><li><a href="#--max-assemble-region-overlap" class="table-of-contents__link"><code>--max-assemble-region-overlap</code></a></li><li><a href="#--assemble-all" class="table-of-contents__link"><code>--assemble-all</code></a></li><li><a href="#--assembler-mask-base-quality" class="table-of-contents__link"><code>--assembler-mask-base-quality</code></a></li><li><a href="#--allow-cycles" class="table-of-contents__link"><code>--allow-cycles</code></a></li><li><a href="#--min-kmer-prune" class="table-of-contents__link"><code>--min-kmer-prune</code></a></li><li><a href="#--max-bubbles" class="table-of-contents__link"><code>--max-bubbles</code></a></li><li><a href="#--min-bubble-score" class="table-of-contents__link"><code>--min-bubble-score</code></a></li><li><a href="#--min-candidate-credible-vaf-probability" class="table-of-contents__link"><code>--min-candidate-credible-vaf-probability</code></a></li></ul></li><li><a href="#haplotype-generation-options" class="table-of-contents__link">Haplotype generation options</a><ul><li><a href="#--max-haplotypes" class="table-of-contents__link"><code>--max-haplotypes</code></a></li><li><a href="#--haplotype-holdout-threshold" class="table-of-contents__link"><code>--haplotype-holdout-threshold</code></a></li><li><a href="#--haplotype-overflow" class="table-of-contents__link"><code>--haplotype-overflow</code></a></li><li><a href="#--max-holdout-depth" class="table-of-contents__link"><code>--max-holdout-depth</code></a></li><li><a href="#--extension-level" class="table-of-contents__link"><code>--extension-level</code></a></li><li><a href="#--lagging-level" class="table-of-contents__link"><code>--lagging-level</code></a></li><li><a href="#--backtrack-level" class="table-of-contents__link"><code>--backtrack-level</code></a></li><li><a href="#--min-protected-haplotype-posterior" class="table-of-contents__link"><code>--min-protected-haplotype-posterior</code></a></li><li><a href="#--dont-protect-reference-haplotype" class="table-of-contents__link"><code>--dont-protect-reference-haplotype</code></a></li><li><a href="#--bad-region-tolerance-1" class="table-of-contents__link"><code>--bad-region-tolerance</code></a></li></ul></li><li><a href="#common-variant-calling-options" class="table-of-contents__link">Common variant calling options</a><ul><li><a href="#--caller" class="table-of-contents__link"><code>--caller</code></a></li><li><a href="#--organism-ploidy" class="table-of-contents__link"><code>--organism-ploidy</code></a></li><li><a href="#--contig-ploidies" class="table-of-contents__link"><code>--contig-ploidies</code></a></li><li><a href="#--contig-ploidies-file" class="table-of-contents__link"><code>--contig-ploidies-file</code></a></li><li><a href="#--min-variant-posterior" class="table-of-contents__link"><code>--min-variant-posterior</code></a></li><li><a href="#--refcall" class="table-of-contents__link"><code>--refcall</code></a></li><li><a href="#--refcall-block-merge-quality" class="table-of-contents__link"><code>--refcall-block-merge-quality</code></a></li><li><a href="#--min-refcall-posterior" class="table-of-contents__link"><code>--min-refcall-posterior</code></a></li><li><a href="#--max-refcall-posterior" class="table-of-contents__link"><code>--max-refcall-posterior</code></a></li><li><a href="#--snp-heterozygosity" class="table-of-contents__link"><code>--snp-heterozygosity</code></a></li><li><a href="#--snp-heterozygosity-stdev" class="table-of-contents__link"><code>--snp-heterozygosity-stdev</code></a></li><li><a href="#--indel-heterozygosity" class="table-of-contents__link"><code>--indel-heterozygosity</code></a></li><li><a href="#--max-genotypes" class="table-of-contents__link"><code>--max-genotypes</code></a></li><li><a href="#--max-genotype-combinations" class="table-of-contents__link"><code>--max-genotype-combinations</code></a></li><li><a href="#--use-uniform-genotype-priors" class="table-of-contents__link"><code>--use-uniform-genotype-priors</code></a></li><li><a href="#--use-independent-genotype-priors" class="table-of-contents__link"><code>--use-independent-genotype-priors</code></a></li><li><a href="#--model-posterior" class="table-of-contents__link"><code>--model-posterior</code></a></li><li><a href="#--disable-inactive-flank-scoring" class="table-of-contents__link"><code>--disable-inactive-flank-scoring</code></a></li><li><a href="#--dont-model-mapping-quality" class="table-of-contents__link"><code>--dont-model-mapping-quality</code></a></li><li><a href="#--sequence-error-model" class="table-of-contents__link"><code>--sequence-error-model</code></a></li><li><a href="#--max-indel-errors" class="table-of-contents__link"><code>--max-indel-errors</code></a></li><li><a href="#--use-wide-hmm-scores" class="table-of-contents__link"><code>--use-wide-hmm-scores</code></a></li><li><a href="#--max-vb-seeds" class="table-of-contents__link"><code>--max-vb-seeds</code></a></li><li><a href="#--read-linkage" class="table-of-contents__link"><code>--read-linkage</code></a></li><li><a href="#--min-phase-score" class="table-of-contents__link"><code>--min-phase-score</code></a></li><li><a href="#--disable-early-phase-detection" class="table-of-contents__link"><code>--disable-early-phase-detection</code></a></li></ul></li><li><a href="#cancer-variant-calling-options" class="table-of-contents__link">Cancer variant calling options</a><ul><li><a href="#--normal-samples" class="table-of-contents__link"><code>--normal-samples</code></a></li><li><a href="#--max-somatic-haplotypes" class="table-of-contents__link"><code>--max-somatic-haplotypes</code></a></li><li><a href="#--somatic-snv-prior" class="table-of-contents__link"><code>--somatic-snv-prior</code></a></li><li><a href="#--somatic-indel-prior" class="table-of-contents__link"><code>--somatic-indel-prior</code></a></li><li><a href="#--min-expected-somatic-frequency" class="table-of-contents__link"><code>--min-expected-somatic-frequency</code></a></li><li><a href="#--min-credible-somatic-frequency" class="table-of-contents__link"><code>--min-credible-somatic-frequency</code></a></li><li><a href="#--tumour-germline-concentration" class="table-of-contents__link"><code>--tumour-germline-concentration</code></a></li><li><a href="#--somatic-credible-mass" class="table-of-contents__link"><code>--somatic-credible-mass</code></a></li><li><a href="#--min-somatic-posterior" class="table-of-contents__link"><code>--min-somatic-posterior</code></a></li><li><a href="#--normal-contamination-risk" class="table-of-contents__link"><code>--normal-contamination-risk</code></a></li><li><a href="#--somatics-only" class="table-of-contents__link"><code>--somatics-only</code></a></li></ul></li><li><a href="#trio-variant-calling-options" class="table-of-contents__link">Trio variant calling options</a><ul><li><a href="#--maternal-sample" class="table-of-contents__link"><code>--maternal-sample</code></a></li><li><a href="#--paternal-sample" class="table-of-contents__link"><code>--paternal-sample</code></a></li><li><a href="#--denovo-snv-prior" class="table-of-contents__link"><code>--denovo-snv-prior</code></a></li><li><a href="#--denovo-indel-prior" class="table-of-contents__link"><code>--denovo-indel-prior</code></a></li><li><a href="#--min-denovo-posterior" class="table-of-contents__link"><code>--min-denovo-posterior</code></a></li><li><a href="#--denovos-only" class="table-of-contents__link"><code>--denovos-only</code></a></li></ul></li><li><a href="#polyclone-variant-calling-options" class="table-of-contents__link">Polyclone variant calling options</a><ul><li><a href="#--max-clones" class="table-of-contents__link"><code>--max-clones</code></a></li><li><a href="#--min-clone-frequency" class="table-of-contents__link"><code>--min-clone-frequency</code></a></li><li><a href="#--clone-prior" class="table-of-contents__link"><code>--clone-prior</code></a></li><li><a href="#--clone-concentration" class="table-of-contents__link"><code>--clone-concentration</code></a></li></ul></li><li><a href="#cell-variant-calling-options" class="table-of-contents__link">Cell variant calling options</a><ul><li><a href="#--max-copy-loss" class="table-of-contents__link"><code>--max-copy-loss</code></a></li><li><a href="#--max-copy-gains" class="table-of-contents__link"><code>--max-copy-gains</code></a></li><li><a href="#--somatic-cnv-prior" class="table-of-contents__link"><code>--somatic-cnv-prior</code></a></li><li><a href="#--dropout-concentration" class="table-of-contents__link"><code>--dropout-concentration</code></a></li><li><a href="#--sample-dropout-concentration" class="table-of-contents__link"><code>--sample-dropout-concentration</code></a></li><li><a href="#--phylogeny-concentration" class="table-of-contents__link"><code>--phylogeny-concentration</code></a></li></ul></li><li><a href="#variant-filtering-options" class="table-of-contents__link">Variant filtering options</a><ul><li><a href="#--disable-call-filtering" class="table-of-contents__link"><code>--disable-call-filtering</code></a></li><li><a href="#--filter-expression" class="table-of-contents__link"><code>--filter-expression</code></a></li><li><a href="#--somatic-filter-expression" class="table-of-contents__link"><code>--somatic-filter-expression</code></a></li><li><a href="#--denovo-filter-expression" class="table-of-contents__link"><code>--denovo-filter-expression</code></a></li><li><a href="#--refcall-filter-expression" class="table-of-contents__link"><code>--refcall-filter-expression</code></a></li><li><a href="#--use-preprocessed-reads-for-filtering" class="table-of-contents__link"><code>--use-preprocessed-reads-for-filtering</code></a></li><li><a href="#--keep-unfiltered-calls" class="table-of-contents__link"><code>--keep-unfiltered-calls</code></a></li><li><a href="#--annotations" class="table-of-contents__link"><code>--annotations</code></a></li><li><a href="#--filter-vcf" class="table-of-contents__link"><code>--filter-vcf</code></a></li><li><a href="#--forest-model" class="table-of-contents__link"><code>--forest-model</code></a></li><li><a href="#--somatic-forest-model" class="table-of-contents__link"><code>--somatic-forest-model</code></a></li><li><a href="#--min-forest-quality" class="table-of-contents__link"><code>--min-forest-quality</code></a></li><li><a href="#--use-germline-forest-for-somatic-normals" class="table-of-contents__link"><code>--use-germline-forest-for-somatic-normals</code></a></li></ul></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><h4 class="footer__title">Docs</h4><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/octopus/docs/installation">Installation</a></li><li class="footer__item"><a class="footer__link-item" href="/octopus/docs/cli">Command Line</a></li><li class="footer__item"><a class="footer__link-item" href="/octopus/docs/tutorials/germline">Case Studies</a></li></ul></div><div class="col footer__col"><h4 class="footer__title">Community</h4><ul class="footer__items"><li class="footer__item"><a href="https://www.biostars.org/tag/octopus" target="_blank" rel="noopener noreferrer" class="footer__link-item">Biostars</a></li><li class="footer__item"><a href="https://gitter.im/octopus-caller/Lobby" target="_blank" rel="noopener noreferrer" class="footer__link-item">Gitter</a></li><li class="footer__item"><a href="https://twitter.com/octopus" target="_blank" rel="noopener noreferrer" class="footer__link-item">Twitter</a></li></ul></div><div class="col footer__col"><h4 class="footer__title">More</h4><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/octopus/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/luntergroup/octopus" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub</a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright Â© 2022 Daniel Cooke. Built with Docusaurus.</div></div></div></footer></div>
<script src="/octopus/assets/js/runtime~main.e4a68a25.js"></script>
<script src="/octopus/assets/js/main.f8c4a271.js"></script>
</body>
</html>