<!doctype html>
<html lang="en" dir="ltr">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.0">
<link rel="alternate" type="application/rss+xml" href="/octopus/blog/rss.xml" title="Octopus Blog RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/octopus/blog/atom.xml" title="Octopus Blog Atom Feed"><title data-react-helmet="true">Cancer | Octopus</title><meta data-react-helmet="true" property="og:url" content="https://luntergroup.github.io/octopus/docs/guides/models/cancer"><meta data-react-helmet="true" name="docusaurus_locale" content="en"><meta data-react-helmet="true" name="docusaurus_version" content="current"><meta data-react-helmet="true" name="docusaurus_tag" content="docs-default-current"><meta data-react-helmet="true" property="og:title" content="Cancer | Octopus"><meta data-react-helmet="true" name="description" content="The cancer calling model is for calling germline variation and somatic mutations in tumours. The model can jointly genotype multiple tumours from the same individual, and make use of a normal sample for improved classification power."><meta data-react-helmet="true" property="og:description" content="The cancer calling model is for calling germline variation and somatic mutations in tumours. The model can jointly genotype multiple tumours from the same individual, and make use of a normal sample for improved classification power."><link data-react-helmet="true" rel="shortcut icon" href="/octopus/img/favicon.ico"><link data-react-helmet="true" rel="canonical" href="https://luntergroup.github.io/octopus/docs/guides/models/cancer"><link data-react-helmet="true" rel="alternate" href="https://luntergroup.github.io/octopus/docs/guides/models/cancer" hreflang="en"><link data-react-helmet="true" rel="alternate" href="https://luntergroup.github.io/octopus/docs/guides/models/cancer" hreflang="x-default"><link rel="stylesheet" href="/octopus/assets/css/styles.0d769f7d.css">
<link rel="preload" href="/octopus/assets/js/runtime~main.cfeb4a38.js" as="script">
<link rel="preload" href="/octopus/assets/js/main.668a8a5c.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div><a href="#main" class="skipToContent_OuoZ">Skip to main content</a></div><nav class="navbar navbar--fixed-top navbarHideable_RReh"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle" type="button" tabindex="0"><svg aria-label="Menu" width="30" height="30" viewBox="0 0 30 30" role="img" focusable="false"><title>Menu</title><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/octopus/"><img src="/octopus/img/octopus.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--light_4Vu1 navbar__logo"><img src="/octopus/img/octopus_keytar.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--dark_uzRr navbar__logo"><strong class="navbar__title">Octopus</strong></a><a class="navbar__item navbar__link navbar__link--active" href="/octopus/docs/introduction">Docs</a><a class="navbar__item navbar__link" href="/octopus/docs/cli">Command line</a><a class="navbar__item navbar__link" href="/octopus/docs/publications">Publications</a><a class="navbar__item navbar__link" href="/octopus/blog">Blog</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/luntergroup/octopus" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="react-toggle displayOnlyInLargeViewport_cxYs react-toggle--disabled" role="button" tabindex="-1"><div class="react-toggle-track"><div class="react-toggle-track-check"><span class="toggle_iYfV">ðŸŒœ</span></div><div class="react-toggle-track-x"><span class="toggle_iYfV">ðŸŒž</span></div></div><div class="react-toggle-thumb"></div><input type="checkbox" class="react-toggle-screenreader-only" aria-label="Switch between dark and light mode"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div><div class="navbar-sidebar"><div class="navbar-sidebar__brand"><a class="navbar__brand" href="/octopus/"><img src="/octopus/img/octopus.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--light_4Vu1 navbar__logo"><img src="/octopus/img/octopus_keytar.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--dark_uzRr navbar__logo"><strong class="navbar__title">Octopus</strong></a></div><div class="navbar-sidebar__items"><div class="menu"><ul class="menu__list"><li class="menu__list-item"><a class="menu__link navbar__link--active" href="/octopus/docs/introduction">Docs</a></li><li class="menu__list-item"><a class="menu__link" href="/octopus/docs/cli">Command line</a></li><li class="menu__list-item"><a class="menu__link" href="/octopus/docs/publications">Publications</a></li><li class="menu__list-item"><a class="menu__link" href="/octopus/blog">Blog</a></li><li class="menu__list-item"><a href="https://github.com/luntergroup/octopus" target="_blank" rel="noopener noreferrer" class="menu__link header-github-link" aria-label="GitHub repository"></a></li></ul></div></div></div></nav><div class="main-wrapper docs-wrapper doc-page"><div class="docPage_lDyR"><div class="docSidebarContainer_0YBq" role="complementary"><div class="sidebar_a3j0 sidebarWithHideableNavbar_VlPv"><a tabindex="-1" class="sidebarLogo_hmkv" href="/octopus/"><img src="/octopus/img/octopus.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--light_4Vu1"><img src="/octopus/img/octopus_keytar.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--dark_uzRr"><strong>Octopus</strong></a><div class="menu menu--responsive thin-scrollbar menu_cyFh"><button aria-label="Open menu" aria-haspopup="true" class="button button--secondary button--sm menu__button" type="button"><svg aria-label="Menu" class="sidebarMenuIcon_iZzd" width="24" height="24" viewBox="0 0 30 30" role="img" focusable="false"><title>Menu</title><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><ul class="menu__list"><li class="menu__list-item"><a class="menu__link menu__link--sublist" href="#!">Getting Started</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/introduction">Introduction</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/installation">Installation</a></li></ul></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#!">Guides</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/overview">Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/preprocessing">Read Preprocessing</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/discovery">Variant Discovery</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/haplotypes">Haplotype Proposal</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#!" tabindex="0">Calling Models</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/models/individual">Individual</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/models/trio">Trio</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/models/population">Population</a></li><li class="menu__list-item"><a aria-current="page" class="menu__link menu__link--active active" tabindex="0" href="/octopus/docs/guides/models/cancer">Cancer</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/models/polyclone">Polyclone</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/models/cell">Cell</a></li></ul></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/errorModels">Error Models</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/phasing">Phasing</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">Filtering</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/filtering/annotations">Introduction</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/filtering/forest">Random Forest</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/filtering/thresholds">Hard Thresholds</a></li></ul></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/bamout">Realigned BAMs</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">Advanced</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/targeted">Targeted Calling</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/threading">Multithreading</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/memory">Memory Use</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/vcf">VCF Format</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/configs">Config files</a></li></ul></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!">Case studies</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/tutorials/germline">Germline WGS</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/tutorials/mitochondria">Mitochondria</a></li></ul></li></ul></div></div></div><main class="docMainContainer_r8cw"><div class="container padding-vert--lg docItemWrapper_NJLN"><div class="row"><div class="col docItemCol_zHA2"><div class="docItemContainer_oiyr"><article><header><h1 class="docTitle_-X99">Cancer</h1></header><div class="markdown"><p>The <code>cancer</code> calling model is for calling germline variation and somatic mutations in tumours. The model can jointly genotype multiple tumours from the same individual, and make use of a normal sample for improved classification power.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="usage-tumour-normal-pairs"></a>Usage: tumour-normal pairs<a class="hash-link" href="#usage-tumour-normal-pairs" title="Direct link to heading">#</a></h2><p>To call germline and somatic mutations in a paired tumour-normal sample, just specify which sample is the normal (<code>--normal-sample</code>; <code>-N</code>):</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R hs37d5.fa -I normal.bam tumour.bam --normal-sample NORMAL</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p>It is also possible to genotype multiple tumours from the same individual jointly:</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R hs37d5.fa -I normal.bam tumourA.bam tumourB.bam -N NORMAL</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="usage-tumour-only"></a>Usage: tumour only<a class="hash-link" href="#usage-tumour-only" title="Direct link to heading">#</a></h2><p>If a normal sample is not present the cancer calling model must be invoked explicitly:</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -R hs37d5.fa -I tumourA.bam tumourB.bam -C cancer</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p>Be aware that without a normal sample, somatic mutation classification power is significantly reduced.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="vcf-output"></a>VCF output<a class="hash-link" href="#vcf-output" title="Direct link to heading">#</a></h2><p>By default both germline and somatic variants are called, somatic mutations are tagged with the <code>SOMATIC</code> INFO field. The <code>GT</code> fields for <code>SOMATIC</code> variants and any other variants in the same phase set (<code>PS</code>) are augmented with the number of unique somatic haplotypes inferred (somatic haplotypes are identified with the <code>HSS</code> tag). For example, in the following VCF:</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI"><div tabindex="0" class="prism-code language-undefined codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">##fileformat=VCFv4.3</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">##INFO=&lt;ID=SOMATIC,Number=0,Type=Flag,Description=&quot;Record includes a somatic mutation&quot;&gt;</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">##FORMAT=&lt;ID=GT,Number=1,Type=String,Description=&quot;Genotype&quot;&gt;</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">##FORMAT=&lt;ID=PS,Number=1,Type=String,Description=&quot;Phase set&quot;&gt;</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">##FORMAT=&lt;ID=HSS,Number=.,Type=Integer,Description=&quot;Somatic status for each haplotype&quot;&gt;</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">#CHROM POS ID REF ALT QUAL FILTER INFO FORMAT NORMAL TUMOUR</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">1 50  . G  A   . . SOMATIC GT:PS:HSS   0|0:50:0,0    1|0|0:50:1,0,0</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">1 100 . A  C   . . SOMATIC GT:PS:HSS   0|0:100:0,0   0|0|0|1:100:0,0,1,1</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">1 150 . G  T   . . .       GT:PS:HSS   1|0:100:0,0   1|0|1|0:100:0,0,1,1</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">1 200 . A  AC  . . SOMATIC GT:PS:HSS   0|0:100:0,0   0|0|1|0:100:0,0,1,1</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">1 250 . TA T   . . .       GT:PS:HSS   1|1:100:0,0   1|1|1|1:100:0,0,1,1</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">2 300 . C  A,T . . SOMATIC GT:PS:HSS   0|2:300:0,0   1|0|2:300:1,0,0</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">2 350 . G  T   . . .       GT:PS:HSS   0|1:300:0,0   0|0|1:300:1,0,0</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">2 400 . T  C   . . SOMATIC GT:PS:HSS   0|0:300:0,0   1|0|0:300:1,0,0</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">3 100 . A  C   . . SOMATIC GT:PS:HSS   0|0:100:0,0   0|0|1|1:100:0,0,1,1</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">3 150 . C  CC  . . SOMATIC GT:PS:HSS   0|0:100:0,0   0|0|1|0:100:0,0,1,1</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">3 200 . G  T   . . .       GT:PS:HSS   0|1:100:0,0   0|1|1|1:100:0,0,1,1</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">4 100 . C  T   . . .       GT:PS:HSS   0|1:100:0,0   0|1|1|1:100:0,1,1,0</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">4 150 . A  G   . . SOMATIC GT:PS:HSS   0|0:100:0,0   0|0|1|0:100:0,1,1,0</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">4 200 . A  G   . . SOMATIC GT:PS:HSS   0|0:100:0,0   0|1|0|0:100:0,1,1,0</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p>The first phase set <code>1:50</code> includes a simple somatic mutation; it is not phased with any other variants. Downstream of this is another phase block starting at <code>1:100</code> that includes 2 germline and 2 somatic variants. The first somatic mutation in this phase set at <code>1:100</code> is phased onto the germline haplotype including the reference allele at the germline variant at <code>1:150</code>. The second somatic mutation in this phase set is phased with the alternate allele of this germline variant. In the third phase set stating at <code>2:300</code> there is a somatic mutation that segregates with a germline variant at the same position, and another somatic mutation which is phased onto the same germline haplotype - the reference. The somatic allele in the multiallelic record is determined by looking at the <code>HSS</code> flags (1 indicates somatic), so &quot;C&gt;A&quot; is the somatic mutation in this case. The third phase set starting at <code>3:100</code> includes 2 somatic mutations phased onto the same germline haplotype, but the first somatic mutation was inferred to segregate with both the germline allele and somatic allele at <code>3:150</code>, suggests linear progression of the C&gt;CC somatic mutation. The last phase set starting at <code>4:100</code> includes 2 somatic mutations phased onto the same germline haplotype, but not with each other.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="qual-vs-pp"></a><code>QUAL</code> vs <code>PP</code><a class="hash-link" href="#qual-vs-pp" title="Direct link to heading">#</a></h2><p>For both paired and tumour-only calling, octopus reports two quality scores for each call (both germline and somatic):</p><ul><li><code>QUAL</code> is the posterior probability the variant is <em>segregating</em> in the sample regardless of somatic classification.</li><li><code>PP</code> (an <code>INFO</code> field) is the posterior probability the variant is segregating <strong>and</strong> classified correctly.</li></ul><p>The difference between <code>QUAL</code> and <code>PP</code> indicates the uncertainty in the calls classification; a call may have high <code>QUAL</code> but low <code>PP</code> if the classification is uncertain (common in tumour-only calling or if the normal coverage is low). <code>PP</code> should always be less than <code>QUAL</code> in theory.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="hpc-map_hf-and-hf_cr"></a><code>HPC</code>, <code>MAP_HF</code> and <code>HF_CR</code><a class="hash-link" href="#hpc-map_hf-and-hf_cr" title="Direct link to heading">#</a></h2><p>Octopus infers a probability distribution over haplotype frequencies, including any somatic haplotypes. For each variant in a phase-block with a <code>SOMATIC</code> variant, three statistics are reported that relate to haplotype frequency (per haplotype).</p><ul><li><code>HPC</code> is the Dirichlet posterior pseudo-count. Note that this count includes the prior count so should not be taken to mean the empirical count.</li><li><code>MAP_HF</code> (<code>FORMAT</code>) is the <a href="https://en.wikipedia.org/wiki/Maximum_a_posteriori_estimation" target="_blank" rel="noopener noreferrer">Maximum a Posteriori</a> haplotype frequency point estimate.</li><li><code>HF_CR</code> (<code>FORMAT</code>) is a <a href="https://en.wikipedia.org/wiki/Credible_interval" target="_blank" rel="noopener noreferrer">credible interval</a> of the haplotypes frequency. The mass of the credible interval is specified by <code>--credible-mass</code>. The credible interval gives you an indication of how certain the MAP_HF estimate is; A very narrow interval means the MAP estimate is very certain, a wide interval means it is uncertain.</li></ul><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="performance-considerations"></a>Performance considerations<a class="hash-link" href="#performance-considerations" title="Direct link to heading">#</a></h2><ul><li>The number of genotypes considered by the model <code>--max-genotype</code> has a significant impact on overall runtime.</li><li>The parameter <code>--max-somatic-haplotypes</code> controls the maximum number of unique segregating somatic haplotypes to be modelled. There must be at-least one somatic haplotype, but adding more can resolve somatic mutations falling on different germline haplotypes or multiple distinct haplotypes due to sub clonal evolution. </li></ul><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="somatics-only"></a>SOMATICs only<a class="hash-link" href="#somatics-only" title="Direct link to heading">#</a></h2><p>To report only <code>SOMATIC</code> calls just add the <code>--somatics-only</code> command. This will only work if calls are beings filtered already (i.e. it will have no effect if <code>-f off</code>). It is generally not recommended to use this option until you are 100% satisfied with your calls, as call filtering will not work correctly if germline variants have been filtered from the VCF; you will not be able to re-filter your SOMATIC-only VCF.</p></div></article><div class="margin-vert--xl"><div class="row"><div class="col"><a href="https://github.com/${organizationName}/${projectName}/edit/${branch}/website/docs/guides/models/cancer.md" target="_blank" rel="noreferrer noopener"><svg fill="currentColor" height="1.2em" width="1.2em" preserveAspectRatio="xMidYMid meet" role="img" viewBox="0 0 40 40" class="iconEdit_mS5F" aria-label="Edit page"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div></div></div><div class="margin-vert--lg"><nav class="pagination-nav" aria-label="Docs pages navigation"><div class="pagination-nav__item"><a class="pagination-nav__link" href="/octopus/docs/guides/models/population"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Â« Population</div></a></div><div class="pagination-nav__item pagination-nav__item--next"><a class="pagination-nav__link" href="/octopus/docs/guides/models/polyclone"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Polyclone Â»</div></a></div></nav></div></div></div><div class="col col--3"><div class="tableOfContents_vrFS thin-scrollbar"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#usage-tumour-normal-pairs" class="table-of-contents__link">Usage: tumour-normal pairs</a></li><li><a href="#usage-tumour-only" class="table-of-contents__link">Usage: tumour only</a></li><li><a href="#vcf-output" class="table-of-contents__link">VCF output</a></li><li><a href="#qual-vs-pp" class="table-of-contents__link"><code>QUAL</code> vs <code>PP</code></a></li><li><a href="#hpc-map_hf-and-hf_cr" class="table-of-contents__link"><code>HPC</code>, <code>MAP_HF</code> and <code>HF_CR</code></a></li><li><a href="#performance-considerations" class="table-of-contents__link">Performance considerations</a></li><li><a href="#somatics-only" class="table-of-contents__link">SOMATICs only</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><h4 class="footer__title">Docs</h4><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/octopus/docs/installation">Installation</a></li><li class="footer__item"><a class="footer__link-item" href="/octopus/docs/cli">Command Line</a></li><li class="footer__item"><a class="footer__link-item" href="/octopus/docs/tutorials/germline">Case Studies</a></li></ul></div><div class="col footer__col"><h4 class="footer__title">Community</h4><ul class="footer__items"><li class="footer__item"><a href="https://www.biostars.org/tag/octopus" target="_blank" rel="noopener noreferrer" class="footer__link-item">Biostars</a></li><li class="footer__item"><a href="https://gitter.im/octopus-caller/Lobby" target="_blank" rel="noopener noreferrer" class="footer__link-item">Gitter</a></li><li class="footer__item"><a href="https://twitter.com/octopus" target="_blank" rel="noopener noreferrer" class="footer__link-item">Twitter</a></li></ul></div><div class="col footer__col"><h4 class="footer__title">More</h4><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/octopus/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/luntergroup/octopus" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub</a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright Â© 2022 Daniel Cooke. Built with Docusaurus.</div></div></div></footer></div>
<script src="/octopus/assets/js/runtime~main.cfeb4a38.js"></script>
<script src="/octopus/assets/js/main.668a8a5c.js"></script>
</body>
</html>