<!doctype html>
<html lang="en" dir="ltr">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.0">
<link rel="alternate" type="application/rss+xml" href="/octopus/blog/rss.xml" title="Octopus Blog RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/octopus/blog/atom.xml" title="Octopus Blog Atom Feed"><title data-react-helmet="true">Cell | Octopus</title><meta data-react-helmet="true" property="og:url" content="https://luntergroup.github.io/octopus/docs/guides/models/cell"><meta data-react-helmet="true" name="docusaurus_locale" content="en"><meta data-react-helmet="true" name="docusaurus_version" content="current"><meta data-react-helmet="true" name="docusaurus_tag" content="docs-default-current"><meta data-react-helmet="true" property="og:title" content="Cell | Octopus"><meta data-react-helmet="true" name="description" content="The cell calling model is used to call germline and somatic variants in single cell and minibatch cell sequencing data. The model attempts to infer local phylogenies for the cells and accounts for allelic biases and dropout often observed in single cell sequencing data."><meta data-react-helmet="true" property="og:description" content="The cell calling model is used to call germline and somatic variants in single cell and minibatch cell sequencing data. The model attempts to infer local phylogenies for the cells and accounts for allelic biases and dropout often observed in single cell sequencing data."><link data-react-helmet="true" rel="shortcut icon" href="/octopus/img/favicon.ico"><link data-react-helmet="true" rel="canonical" href="https://luntergroup.github.io/octopus/docs/guides/models/cell"><link data-react-helmet="true" rel="alternate" href="https://luntergroup.github.io/octopus/docs/guides/models/cell" hreflang="en"><link data-react-helmet="true" rel="alternate" href="https://luntergroup.github.io/octopus/docs/guides/models/cell" hreflang="x-default"><link rel="stylesheet" href="/octopus/assets/css/styles.0d769f7d.css">
<link rel="preload" href="/octopus/assets/js/runtime~main.82d64c00.js" as="script">
<link rel="preload" href="/octopus/assets/js/main.668a8a5c.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div><a href="#main" class="skipToContent_OuoZ">Skip to main content</a></div><nav class="navbar navbar--fixed-top navbarHideable_RReh"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle" type="button" tabindex="0"><svg aria-label="Menu" width="30" height="30" viewBox="0 0 30 30" role="img" focusable="false"><title>Menu</title><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/octopus/"><img src="/octopus/img/octopus.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--light_4Vu1 navbar__logo"><img src="/octopus/img/octopus_keytar.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--dark_uzRr navbar__logo"><strong class="navbar__title">Octopus</strong></a><a class="navbar__item navbar__link navbar__link--active" href="/octopus/docs/introduction">Docs</a><a class="navbar__item navbar__link" href="/octopus/docs/cli">Command line</a><a class="navbar__item navbar__link" href="/octopus/docs/publications">Publications</a><a class="navbar__item navbar__link" href="/octopus/blog">Blog</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/luntergroup/octopus" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="react-toggle displayOnlyInLargeViewport_cxYs react-toggle--disabled" role="button" tabindex="-1"><div class="react-toggle-track"><div class="react-toggle-track-check"><span class="toggle_iYfV">ðŸŒœ</span></div><div class="react-toggle-track-x"><span class="toggle_iYfV">ðŸŒž</span></div></div><div class="react-toggle-thumb"></div><input type="checkbox" class="react-toggle-screenreader-only" aria-label="Switch between dark and light mode"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div><div class="navbar-sidebar"><div class="navbar-sidebar__brand"><a class="navbar__brand" href="/octopus/"><img src="/octopus/img/octopus.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--light_4Vu1 navbar__logo"><img src="/octopus/img/octopus_keytar.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--dark_uzRr navbar__logo"><strong class="navbar__title">Octopus</strong></a></div><div class="navbar-sidebar__items"><div class="menu"><ul class="menu__list"><li class="menu__list-item"><a class="menu__link navbar__link--active" href="/octopus/docs/introduction">Docs</a></li><li class="menu__list-item"><a class="menu__link" href="/octopus/docs/cli">Command line</a></li><li class="menu__list-item"><a class="menu__link" href="/octopus/docs/publications">Publications</a></li><li class="menu__list-item"><a class="menu__link" href="/octopus/blog">Blog</a></li><li class="menu__list-item"><a href="https://github.com/luntergroup/octopus" target="_blank" rel="noopener noreferrer" class="menu__link header-github-link" aria-label="GitHub repository"></a></li></ul></div></div></div></nav><div class="main-wrapper docs-wrapper doc-page"><div class="docPage_lDyR"><div class="docSidebarContainer_0YBq" role="complementary"><div class="sidebar_a3j0 sidebarWithHideableNavbar_VlPv"><a tabindex="-1" class="sidebarLogo_hmkv" href="/octopus/"><img src="/octopus/img/octopus.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--light_4Vu1"><img src="/octopus/img/octopus_keytar.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--dark_uzRr"><strong>Octopus</strong></a><div class="menu menu--responsive thin-scrollbar menu_cyFh"><button aria-label="Open menu" aria-haspopup="true" class="button button--secondary button--sm menu__button" type="button"><svg aria-label="Menu" class="sidebarMenuIcon_iZzd" width="24" height="24" viewBox="0 0 30 30" role="img" focusable="false"><title>Menu</title><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><ul class="menu__list"><li class="menu__list-item"><a class="menu__link menu__link--sublist" href="#!">Getting Started</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/introduction">Introduction</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/installation">Installation</a></li></ul></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#!">Guides</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/overview">Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/preprocessing">Read Preprocessing</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/discovery">Variant Discovery</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/haplotypes">Haplotype Proposal</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#!" tabindex="0">Calling Models</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/models/individual">Individual</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/models/trio">Trio</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/models/population">Population</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/models/cancer">Cancer</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/models/polyclone">Polyclone</a></li><li class="menu__list-item"><a aria-current="page" class="menu__link menu__link--active active" tabindex="0" href="/octopus/docs/guides/models/cell">Cell</a></li></ul></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/errorModels">Error Models</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/phasing">Phasing</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">Filtering</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/filtering/annotations">Introduction</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/filtering/forest">Random Forest</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/filtering/thresholds">Hard Thresholds</a></li></ul></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/bamout">Realigned BAMs</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">Advanced</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/targeted">Targeted Calling</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/threading">Multithreading</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/memory">Memory Use</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/vcf">VCF Format</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/configs">Config files</a></li></ul></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!">Case studies</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/tutorials/germline">Germline WGS</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/tutorials/mitochondria">Mitochondria</a></li></ul></li></ul></div></div></div><main class="docMainContainer_r8cw"><div class="container padding-vert--lg docItemWrapper_NJLN"><div class="row"><div class="col docItemCol_zHA2"><div class="docItemContainer_oiyr"><article><header><h1 class="docTitle_-X99">Cell</h1></header><div class="markdown"><p>The <code>cell</code> calling model is used to call germline and somatic variants in single cell and minibatch cell sequencing data. The model attempts to infer local phylogenies for the cells and accounts for allelic biases and dropout often observed in single cell sequencing data.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="usage-basic"></a>Usage: basic<a class="hash-link" href="#usage-basic" title="Direct link to heading">#</a></h2><p>If all of the samples are single cells and none are control cells:</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -C cell </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    -R ref.fa </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    -I cell1.bam cell2.bam </span><span class="token punctuation" style="color:rgb(199, 146, 234)">..</span><span class="token plain">. cellN.bam </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    -o cells.vcf</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="usage-with-controls"></a>Usage: with controls<a class="hash-link" href="#usage-with-controls" title="Direct link to heading">#</a></h2><p>If the experiment includes control cells (e.g. for tumour-normals) then provide the control cell sample names (<code>--normal-samples</code>; <code>-N</code>):</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -C cell </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    -R ref.fa </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    -I cell1.bam cell2.bam </span><span class="token punctuation" style="color:rgb(199, 146, 234)">..</span><span class="token plain">. cellN.bam </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    --normal-samples CONTROL1 CONTROL2 </span><span class="token punctuation" style="color:rgb(199, 146, 234)">..</span><span class="token plain">. CONTROLM </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    -o cells.vcf</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p>All normal cells are assumed to originate from the root (i.e. founder) node of the phylogeny relating cells, and are therefore assumed to all have the same genotype.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="usage-with-minibatchs"></a>Usage: with minibatchs<a class="hash-link" href="#usage-with-minibatchs" title="Direct link to heading">#</a></h2><p>If any of the samples are derived from minibatches of cells then specify high dropout concentrations (<code>--sample-dropout-concentration</code>) for these samples:</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -C cell </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    -R ref.fa </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    -I cell1.bam cell2.bam </span><span class="token punctuation" style="color:rgb(199, 146, 234)">..</span><span class="token plain">. cellN.bam </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    --sample-dropout-concentration </span><span class="token assign-left variable" style="color:rgb(191, 199, 213)">MINIBATCH1</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token number" style="color:rgb(247, 140, 108)">100</span><span class="token plain"> </span><span class="token assign-left variable" style="color:rgb(191, 199, 213)">MINIBATCH2</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token number" style="color:rgb(247, 140, 108)">100</span><span class="token plain"> </span><span class="token punctuation" style="color:rgb(199, 146, 234)">..</span><span class="token plain"> </span><span class="token assign-left variable" style="color:rgb(191, 199, 213)">MINIBATCHM</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token number" style="color:rgb(247, 140, 108)">100</span><span class="token plain"> </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    -o cells.vcf</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p>The argument for each minibatch sample may reflect the number of cells contained in the minibatch; the larger the number of cells, the larger the argument value.</p><p>The usual use for minibatch samples is for better controls, in which case the minibatches will be normal samples:</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -C cell </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    -R ref.fa </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    -I cell1.bam cell2.bam </span><span class="token punctuation" style="color:rgb(199, 146, 234)">..</span><span class="token plain">. cellN.bam </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    --normal-samples MINIBATCH1 MINIBATCH2 </span><span class="token punctuation" style="color:rgb(199, 146, 234)">..</span><span class="token plain">. MINIBATCHM </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    --sample-dropout-concentration </span><span class="token assign-left variable" style="color:rgb(191, 199, 213)">MINIBATCH1</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token number" style="color:rgb(247, 140, 108)">100</span><span class="token plain"> </span><span class="token assign-left variable" style="color:rgb(191, 199, 213)">MINIBATCH2</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token number" style="color:rgb(247, 140, 108)">100</span><span class="token plain"> </span><span class="token punctuation" style="color:rgb(199, 146, 234)">..</span><span class="token plain"> </span><span class="token assign-left variable" style="color:rgb(191, 199, 213)">MINIBATCHM</span><span class="token operator" style="color:rgb(137, 221, 255)">=</span><span class="token number" style="color:rgb(247, 140, 108)">100</span><span class="token plain"> </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    -o cells.vcf</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h4><a aria-hidden="true" tabindex="-1" class="anchor" id="vcf-output"></a>VCF output<a class="hash-link" href="#vcf-output" title="Direct link to heading">#</a></h4><p>There are several annotations included in the VCF output:</p><table><thead><tr><th>Name</th><th align="center">INFO/FORMAT</th><th align="left">Description</th></tr></thead><tbody><tr><td>SOMATIC</td><td align="center">INFO</td><td align="left">Indicates that a somatic mutation was inferred (i.e. the phylogeny contains more than one node).</td></tr><tr><td>PY</td><td align="center">INFO</td><td align="left">The MAP phylogeny inferred for the variant loci. This annotation is only added for SOMATIC calls.</td></tr><tr><td>PPP</td><td align="center">INFO</td><td align="left">Posterior probability (Phred) for the MAP phylogeny.</td></tr><tr><td>PSPP</td><td align="center">INFO</td><td align="left">Posterior probabilities (Phred) that the local phylogeny contains <code>0</code>,<code>1</code>,... nodes</td></tr><tr><td>PNAP</td><td align="center">FORMAT</td><td align="left">Posterior probabilities (Phred) that this sample is assigned to node ID <code>0</code>,<code>1</code>,.. in the MAP phylogeny (<code>PY</code>).</td></tr></tbody></table><h5><a aria-hidden="true" tabindex="-1" class="anchor" id="py-notation"></a><code>PY</code> notation<a class="hash-link" href="#py-notation" title="Direct link to heading">#</a></h5><p>The phylogeny is serialised using the following algorithm:</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI"><div tabindex="0" class="prism-code language-undefined codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">def serialise(result, node):</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    if (node != NULL):</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">        result += &#x27;(&#x27; + str(node.id)</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">        for child in node:</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">            serialise(result, child)</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">        result += &#x27;)&#x27;</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p>The algorithm is called with the root node of the phylogeny <code>serialise(&quot;&quot;, ROOT)</code>. Examples:</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI"><div tabindex="0" class="prism-code language-undefined codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">     0</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">   /   \</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">  1     2</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain" style="display:inline-block">
</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">(0(1)(2))</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI"><div tabindex="0" class="prism-code language-undefined codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">     0</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">       \</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">        1</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">         \</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">          2</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain" style="display:inline-block">
</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">(0(1(2)))</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><h4><a aria-hidden="true" tabindex="-1" class="anchor" id="cnv-calling"></a>CNV calling<a class="hash-link" href="#cnv-calling" title="Direct link to heading">#</a></h4><p>The model can try to identify local copy changes (i.e. deletions or gains of haplotypes). This will result in some samples having called genotypes with different ploidies to the default ploidy. The maximum number of gains and losses is specified with the <code>--max-copy-gain</code> and <code>--max-copy-loss</code> options, respectively. For example, to identify up to one copy gain or loss:</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus -C cell </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    -R ref.fa </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    -I cell1.bam cell2.bam </span><span class="token punctuation" style="color:rgb(199, 146, 234)">..</span><span class="token plain">. cellN.bam </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    --max-copy-gain </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain"> --max-copy-loss </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain"> </span><span class="token punctuation" style="color:rgb(199, 146, 234)">\</span><span class="token plain"></span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">    -o cells.vcf</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p><strong>Warning</strong> calling copy gains is currently computationally very expensive.</p><h4><a aria-hidden="true" tabindex="-1" class="anchor" id="performance-considerations"></a>Performance considerations<a class="hash-link" href="#performance-considerations" title="Direct link to heading">#</a></h4><p>A critical parameter for this calling model is the maximum size of the phylogeny (<code>--max-clones</code>). Copy loss and gain calling are also computationally expensive.</p><p>It is recommended to allow automatic thread usage with this calling model (use <code>--threads</code> option without an argument).</p></div></article><div class="margin-vert--xl"><div class="row"><div class="col"><a href="https://github.com/${organizationName}/${projectName}/edit/${branch}/website/docs/guides/models/cell.md" target="_blank" rel="noreferrer noopener"><svg fill="currentColor" height="1.2em" width="1.2em" preserveAspectRatio="xMidYMid meet" role="img" viewBox="0 0 40 40" class="iconEdit_mS5F" aria-label="Edit page"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div></div></div><div class="margin-vert--lg"><nav class="pagination-nav" aria-label="Docs pages navigation"><div class="pagination-nav__item"><a class="pagination-nav__link" href="/octopus/docs/guides/models/polyclone"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Â« Polyclone</div></a></div><div class="pagination-nav__item pagination-nav__item--next"><a class="pagination-nav__link" href="/octopus/docs/guides/errorModels"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Error Models Â»</div></a></div></nav></div></div></div><div class="col col--3"><div class="tableOfContents_vrFS thin-scrollbar"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#usage-basic" class="table-of-contents__link">Usage: basic</a></li><li><a href="#usage-with-controls" class="table-of-contents__link">Usage: with controls</a></li><li><a href="#usage-with-minibatchs" class="table-of-contents__link">Usage: with minibatchs</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><h4 class="footer__title">Docs</h4><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/octopus/docs/installation">Installation</a></li><li class="footer__item"><a class="footer__link-item" href="/octopus/docs/cli">Command Line</a></li><li class="footer__item"><a class="footer__link-item" href="/octopus/docs/tutorials/germline">Case Studies</a></li></ul></div><div class="col footer__col"><h4 class="footer__title">Community</h4><ul class="footer__items"><li class="footer__item"><a href="https://www.biostars.org/tag/octopus" target="_blank" rel="noopener noreferrer" class="footer__link-item">Biostars</a></li><li class="footer__item"><a href="https://gitter.im/octopus-caller/Lobby" target="_blank" rel="noopener noreferrer" class="footer__link-item">Gitter</a></li><li class="footer__item"><a href="https://twitter.com/octopus" target="_blank" rel="noopener noreferrer" class="footer__link-item">Twitter</a></li></ul></div><div class="col footer__col"><h4 class="footer__title">More</h4><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/octopus/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/luntergroup/octopus" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub</a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright Â© 2022 Daniel Cooke. Built with Docusaurus.</div></div></div></footer></div>
<script src="/octopus/assets/js/runtime~main.82d64c00.js"></script>
<script src="/octopus/assets/js/main.668a8a5c.js"></script>
</body>
</html>