<!doctype html>
<html lang="en" dir="ltr">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.0">
<link rel="alternate" type="application/rss+xml" href="/octopus/blog/rss.xml" title="Octopus Blog RSS Feed">
<link rel="alternate" type="application/atom+xml" href="/octopus/blog/atom.xml" title="Octopus Blog Atom Feed"><title data-react-helmet="true">Introduction | Octopus</title><meta data-react-helmet="true" property="og:url" content="https://luntergroup.github.io/octopus/docs/guides/filtering/annotations"><meta data-react-helmet="true" name="docusaurus_locale" content="en"><meta data-react-helmet="true" name="docusaurus_version" content="current"><meta data-react-helmet="true" name="docusaurus_tag" content="docs-default-current"><meta data-react-helmet="true" property="og:title" content="Introduction | Octopus"><meta data-react-helmet="true" name="description" content="Octopus calls variants and genotypes using a Bayesian model. Like any model, there are some aspects of real error that are not fully captured in the model which can lead to false calls. We therefore recommended filtering calls to improve precision."><meta data-react-helmet="true" property="og:description" content="Octopus calls variants and genotypes using a Bayesian model. Like any model, there are some aspects of real error that are not fully captured in the model which can lead to false calls. We therefore recommended filtering calls to improve precision."><link data-react-helmet="true" rel="shortcut icon" href="/octopus/img/favicon.ico"><link data-react-helmet="true" rel="canonical" href="https://luntergroup.github.io/octopus/docs/guides/filtering/annotations"><link data-react-helmet="true" rel="alternate" href="https://luntergroup.github.io/octopus/docs/guides/filtering/annotations" hreflang="en"><link data-react-helmet="true" rel="alternate" href="https://luntergroup.github.io/octopus/docs/guides/filtering/annotations" hreflang="x-default"><link rel="stylesheet" href="/octopus/assets/css/styles.0d769f7d.css">
<link rel="preload" href="/octopus/assets/js/runtime~main.4e3f0b2b.js" as="script">
<link rel="preload" href="/octopus/assets/js/main.f8c4a271.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div><a href="#main" class="skipToContent_OuoZ">Skip to main content</a></div><nav class="navbar navbar--fixed-top navbarHideable_RReh"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle" type="button" tabindex="0"><svg aria-label="Menu" width="30" height="30" viewBox="0 0 30 30" role="img" focusable="false"><title>Menu</title><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/octopus/"><img src="/octopus/img/octopus.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--light_4Vu1 navbar__logo"><img src="/octopus/img/octopus_keytar.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--dark_uzRr navbar__logo"><strong class="navbar__title">Octopus</strong></a><a class="navbar__item navbar__link navbar__link--active" href="/octopus/docs/introduction">Docs</a><a class="navbar__item navbar__link" href="/octopus/docs/cli">Command line</a><a class="navbar__item navbar__link" href="/octopus/docs/publications">Publications</a><a class="navbar__item navbar__link" href="/octopus/blog">Blog</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/luntergroup/octopus" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link header-github-link" aria-label="GitHub repository"></a><div class="react-toggle displayOnlyInLargeViewport_cxYs react-toggle--disabled" role="button" tabindex="-1"><div class="react-toggle-track"><div class="react-toggle-track-check"><span class="toggle_iYfV">ðŸŒœ</span></div><div class="react-toggle-track-x"><span class="toggle_iYfV">ðŸŒž</span></div></div><div class="react-toggle-thumb"></div><input type="checkbox" class="react-toggle-screenreader-only" aria-label="Switch between dark and light mode"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div><div class="navbar-sidebar"><div class="navbar-sidebar__brand"><a class="navbar__brand" href="/octopus/"><img src="/octopus/img/octopus.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--light_4Vu1 navbar__logo"><img src="/octopus/img/octopus_keytar.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--dark_uzRr navbar__logo"><strong class="navbar__title">Octopus</strong></a></div><div class="navbar-sidebar__items"><div class="menu"><ul class="menu__list"><li class="menu__list-item"><a class="menu__link navbar__link--active" href="/octopus/docs/introduction">Docs</a></li><li class="menu__list-item"><a class="menu__link" href="/octopus/docs/cli">Command line</a></li><li class="menu__list-item"><a class="menu__link" href="/octopus/docs/publications">Publications</a></li><li class="menu__list-item"><a class="menu__link" href="/octopus/blog">Blog</a></li><li class="menu__list-item"><a href="https://github.com/luntergroup/octopus" target="_blank" rel="noopener noreferrer" class="menu__link header-github-link" aria-label="GitHub repository"></a></li></ul></div></div></div></nav><div class="main-wrapper docs-wrapper doc-page"><div class="docPage_lDyR"><div class="docSidebarContainer_0YBq" role="complementary"><div class="sidebar_a3j0 sidebarWithHideableNavbar_VlPv"><a tabindex="-1" class="sidebarLogo_hmkv" href="/octopus/"><img src="/octopus/img/octopus.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--light_4Vu1"><img src="/octopus/img/octopus_keytar.svg" alt="Octopus Logo" class="themedImage_TMUO themedImage--dark_uzRr"><strong>Octopus</strong></a><div class="menu menu--responsive thin-scrollbar menu_cyFh"><button aria-label="Open menu" aria-haspopup="true" class="button button--secondary button--sm menu__button" type="button"><svg aria-label="Menu" class="sidebarMenuIcon_iZzd" width="24" height="24" viewBox="0 0 30 30" role="img" focusable="false"><title>Menu</title><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><ul class="menu__list"><li class="menu__list-item"><a class="menu__link menu__link--sublist" href="#!">Getting Started</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/introduction">Introduction</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/installation">Installation</a></li></ul></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#!">Guides</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/overview">Overview</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/preprocessing">Read Preprocessing</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/discovery">Variant Discovery</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/haplotypes">Haplotype Proposal</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">Calling Models</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/models/individual">Individual</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/models/trio">Trio</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/models/population">Population</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/models/cancer">Cancer</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/models/polyclone">Polyclone</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/models/cell">Cell</a></li></ul></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/errorModels">Error Models</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/phasing">Phasing</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#!" tabindex="0">Filtering</a><ul class="menu__list"><li class="menu__list-item"><a aria-current="page" class="menu__link menu__link--active active" tabindex="0" href="/octopus/docs/guides/filtering/annotations">Introduction</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/filtering/forest">Random Forest</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/filtering/thresholds">Hard Thresholds</a></li></ul></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/octopus/docs/guides/bamout">Realigned BAMs</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">Advanced</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/targeted">Targeted Calling</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/threading">Multithreading</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/memory">Memory Use</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/vcf">VCF Format</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/guides/advanced/configs">Config files</a></li></ul></li></ul></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!">Case studies</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/octopus/docs/tutorials/germline">Germline WGS</a></li></ul></li></ul></div></div></div><main class="docMainContainer_r8cw"><div class="container padding-vert--lg docItemWrapper_NJLN"><div class="row"><div class="col docItemCol_zHA2"><div class="docItemContainer_oiyr"><article><header><h1 class="docTitle_-X99">Introduction</h1></header><div class="markdown"><p>Octopus calls variants and genotypes using a Bayesian model. Like any model, there are some aspects of real error that are not fully captured in the model which can lead to false calls. We therefore recommended filtering calls to improve precision.</p><p>There are currently two approaches available in Octopus for filtering variants:</p><ol><li><a href="/octopus/docs/guides/filtering/thresholds">Hard coded thresholds</a></li><li><a href="/octopus/docs/guides/filtering/forest">Random forests</a></li></ol><p>Both methods use <a href="#annotations">annotations</a> computed by Octopus.</p><p>The random forest approach is preferred when sufficient training data is available (e.g. typical germline and somatic calling). Hard coded thresholds are appropriate for other cases (e.g. UMI low frequency calling).</p><h2><a aria-hidden="true" tabindex="-1" class="anchor" id="annotations"></a>Annotations<a class="hash-link" href="#annotations" title="Direct link to heading">#</a></h2><p>Octopus provides various annotations for filtering variants. To list available annotations, use the command </p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">$ octopus --help --annotations</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div><p>For example, the current annotations are:</p><div class="codeBlockContainer_J+bg"><div class="codeBlockContent_csEI shell"><div tabindex="0" class="prism-code language-shell codeBlock_rtdJ thin-scrollbar"><div class="codeBlockLines_1zSZ" style="color:#bfc7d5;background-color:#292d3e"><div class="token-line" style="color:#bfc7d5"><span class="token plain">Name    Kind    Number  Type    Description</span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">AC  FORMAT  A   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Number of non-reference alleles called for each sample&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">AD  FORMAT  R   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Empirical allele depth&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">ADP FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Number of reads overlapping the position that could be assigned to an allele&quot;</span><span class="token plain">  </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">AF  FORMAT  R   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Empirical allele frequency (AD / ADP)&quot;</span><span class="token plain"> </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">AFB FORMAT  R   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Absolute difference between empirical allele frequency (AF) and expected allele frequency given genotype&quot;</span><span class="token plain">  </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">AMQ FORMAT  R   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Median mapping quality of reads assigned to each allele&quot;</span><span class="token plain">   </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">ARF FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Fraction of reads overlapping the call that cannot be assigned to a unique haplotype&quot;</span><span class="token plain">  </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">BMC FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Number of base mismatches at variant position in reads supporting variant haplotype&quot;</span><span class="token plain">   </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">BMF FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Fraction of base mismatches at variant position&quot;</span><span class="token plain">   </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">BMQ FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Median quality of base mismatches in reads assigned to a unique allele&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">BQ  FORMAT  R   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Median base quality of reads supporting each allele&quot;</span><span class="token plain">   </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">CC  INFO    </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;PP divided by QUAL&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">CRF INFO    </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Fraction of clipped reads covering the call&quot;</span><span class="token plain">   </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">DAD FORMAT  R   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Number of realigned reads supporting ALT alleles identified as duplicates&quot;</span><span class="token plain"> </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">DAF FORMAT  R   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Fraction of realigned reads supporting ALT alleles identified as duplicates&quot;</span><span class="token plain">   </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">DC  INFO    </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Number of reads supporting a de novo haplotype in the normal&quot;</span><span class="token plain">  </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">DENOVO  FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;DENOVO status of each sample&quot;</span><span class="token plain">  </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">DP  FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Number of read overlapping the call&quot;</span><span class="token plain">   </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">DPC FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Concordance of allele support from duplicate reads&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">ER  FORMAT  R   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Error rate in supporting reads&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">ERS FORMAT  R   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Error rate standard deviation in supporting reads&quot;</span><span class="token plain"> </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">FRF FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Fraction of reads filtered for calling&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">GC  INFO    </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;GC bias of the reference in a window centred on the call&quot;</span><span class="token plain">  </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">GQ  INFO    A   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Number of non-reference alleles called for each sample&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">GQD FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;GQ divided by DP&quot;</span><span class="token plain">  </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">ITV INFO    A   Flag    </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Is the variant a transversion&quot;</span><span class="token plain"> </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">MC  FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Number of mismatches at variant position in reads supporting variant haplotype&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">MF  FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Fraction of reads with mismatches at variant position&quot;</span><span class="token plain"> </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">MHL FORMAT  </span><span class="token builtin class-name" style="color:rgb(255, 203, 107)">.</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Mean likelihood (Phreds) of reads overlapping the site assigned to each haplotype&quot;</span><span class="token plain"> </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">MP  INFO    </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Model posterior for this haplotype block&quot;</span><span class="token plain">  </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">MQ  INFO    </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Mean mapping quality of reads overlapping the call&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">MQ0 INFO    </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Number of reads overlapping the call with mapping quality zero&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">MQD FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Maximum pairwise difference in median mapping qualities of reads supporting each haplotype&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">MRC FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Number of reads supporting the call that appear misaligned&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">MRL FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Maximum read length overlapping the site&quot;</span><span class="token plain">  </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">NC  INFO    </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Fraction of overlapping reads supporting a somatic haplotype in the normal&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">PLN FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Length of the phase block for the call&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">PP  INFO    </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Call posterior probability&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">PPD INFO    </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;PP divided by DP&quot;</span><span class="token plain">  </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">QD  INFO    </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;QUAL divided by DP&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">QUAL    INFO    A   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Number of non-reference alleles called for each sample&quot;</span><span class="token plain">    </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">REB FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Probability allele occurs at the end (head or tail) of supporting reads&quot;</span><span class="token plain">   </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">REFCALL FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;REFCALL status of each sample&quot;</span><span class="token plain"> </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">RSB FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Bias of variant side (head or tail half) in supporting reads&quot;</span><span class="token plain">  </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">RTB FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Probability allele occurs in the tail of supporting reads&quot;</span><span class="token plain"> </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">SB  FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Strand bias of reads based on haplotype support&quot;</span><span class="token plain">   </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">SD  FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Strand bias of reads overlapping the site; probability mass in tails of Beta distribution&quot;</span><span class="token plain"> </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">SF  FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Float   </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Max fraction of reads supporting ALT alleles that are supplementary&quot;</span><span class="token plain">   </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">SHC FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Number of called somatic haplotypes&quot;</span><span class="token plain">   </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">SMQ FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Median mapping quality of reads assigned to called somatic haplotypes&quot;</span><span class="token plain"> </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">SOMATIC FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;SOMATIC status of each sample&quot;</span><span class="token plain"> </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">STRL    INFO    </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Length of overlapping STR&quot;</span><span class="token plain"> </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">STRP    INFO    </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Period of overlapping STR&quot;</span><span class="token plain"> </span></div><div class="token-line" style="color:#bfc7d5"><span class="token plain">VL  FORMAT  </span><span class="token number" style="color:rgb(247, 140, 108)">1</span><span class="token plain">   Integer </span><span class="token string" style="color:rgb(195, 232, 141)">&quot;Maximum length of called alleles&quot;</span></div></div></div><button type="button" aria-label="Copy code to clipboard" class="copyButton_M3SB">Copy</button></div></div></div></article><div class="margin-vert--xl"><div class="row"><div class="col"><a href="https://github.com/${organizationName}/${projectName}/edit/${branch}/website/docs/guides/filtering/annotations.md" target="_blank" rel="noreferrer noopener"><svg fill="currentColor" height="1.2em" width="1.2em" preserveAspectRatio="xMidYMid meet" role="img" viewBox="0 0 40 40" class="iconEdit_mS5F" aria-label="Edit page"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div></div></div><div class="margin-vert--lg"><nav class="pagination-nav" aria-label="Docs pages navigation"><div class="pagination-nav__item"><a class="pagination-nav__link" href="/octopus/docs/guides/phasing"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Â« Phasing</div></a></div><div class="pagination-nav__item pagination-nav__item--next"><a class="pagination-nav__link" href="/octopus/docs/guides/filtering/forest"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Random Forest Â»</div></a></div></nav></div></div></div><div class="col col--3"><div class="tableOfContents_vrFS thin-scrollbar"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#annotations" class="table-of-contents__link">Annotations</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><h4 class="footer__title">Docs</h4><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/octopus/docs/installation">Installation</a></li><li class="footer__item"><a class="footer__link-item" href="/octopus/docs/cli">Command Line</a></li><li class="footer__item"><a class="footer__link-item" href="/octopus/docs/tutorials/germline">Case Studies</a></li></ul></div><div class="col footer__col"><h4 class="footer__title">Community</h4><ul class="footer__items"><li class="footer__item"><a href="https://www.biostars.org/tag/octopus" target="_blank" rel="noopener noreferrer" class="footer__link-item">Biostars</a></li><li class="footer__item"><a href="https://gitter.im/octopus-caller/Lobby" target="_blank" rel="noopener noreferrer" class="footer__link-item">Gitter</a></li><li class="footer__item"><a href="https://twitter.com/octopus" target="_blank" rel="noopener noreferrer" class="footer__link-item">Twitter</a></li></ul></div><div class="col footer__col"><h4 class="footer__title">More</h4><ul class="footer__items"><li class="footer__item"><a class="footer__link-item" href="/octopus/blog">Blog</a></li><li class="footer__item"><a href="https://github.com/luntergroup/octopus" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub</a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright Â© 2022 Daniel Cooke. Built with Docusaurus.</div></div></div></footer></div>
<script src="/octopus/assets/js/runtime~main.4e3f0b2b.js"></script>
<script src="/octopus/assets/js/main.f8c4a271.js"></script>
</body>
</html>